{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Initialization\" data-toc-modified-id=\"Initialization-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Initialization</a></span></li><li><span><a href=\"#Change-in-motif-information-with-ChIP-seq-and-biological-parameters-(Figure-2)\" data-toc-modified-id=\"Change-in-motif-information-with-ChIP-seq-and-biological-parameters-(Figure-2)-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Change in motif information with ChIP-seq and biological parameters (Figure 2)</a></span><ul class=\"toc-item\"><li><span><a href=\"#Initialization-for-Figure-2\" data-toc-modified-id=\"Initialization-for-Figure-2-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Initialization for Figure 2</a></span></li><li><span><a href=\"#Motif-information-and-fidelity-content-decreases-with-increasing-TF-chemical-potential-[Supplement]\" data-toc-modified-id=\"Motif-information-and-fidelity-content-decreases-with-increasing-TF-chemical-potential-[Supplement]-2.2\"><span class=\"toc-item-num\">2.2&nbsp;&nbsp;</span>Motif information and fidelity content decreases with increasing TF chemical potential [Supplement]</a></span></li><li><span><a href=\"#Figure-2A\" data-toc-modified-id=\"Figure-2A-2.3\"><span class=\"toc-item-num\">2.3&nbsp;&nbsp;</span>Figure 2A</a></span></li><li><span><a href=\"#Figure-2B-[PCR-amplification]\" data-toc-modified-id=\"Figure-2B-[PCR-amplification]-2.4\"><span class=\"toc-item-num\">2.4&nbsp;&nbsp;</span>Figure 2B [PCR amplification]</a></span></li><li><span><a href=\"#Figure-2B-[Extraction-efficiency]\" data-toc-modified-id=\"Figure-2B-[Extraction-efficiency]-2.5\"><span class=\"toc-item-num\">2.5&nbsp;&nbsp;</span>Figure 2B [Extraction efficiency]</a></span></li><li><span><a href=\"#Figure-3A-[Top,-sample-output,-indirect-binding]\" data-toc-modified-id=\"Figure-3A-[Top,-sample-output,-indirect-binding]-2.6\"><span class=\"toc-item-num\">2.6&nbsp;&nbsp;</span>Figure 3A [Top, sample output, indirect binding]</a></span></li><li><span><a href=\"#Figure-3A-[Bottom,-sample-output,-cooperative-binding]\" data-toc-modified-id=\"Figure-3A-[Bottom,-sample-output,-cooperative-binding]-2.7\"><span class=\"toc-item-num\">2.7&nbsp;&nbsp;</span>Figure 3A [Bottom, sample output, cooperative binding]</a></span></li><li><span><a href=\"#Figure-3B-[Middle-Row,-Left,-Indirect-binding]\" data-toc-modified-id=\"Figure-3B-[Middle-Row,-Left,-Indirect-binding]-2.8\"><span class=\"toc-item-num\">2.8&nbsp;&nbsp;</span>Figure 3B [Middle Row, Left, Indirect binding]</a></span></li><li><span><a href=\"#Figure-3B-[Cooperatively-bound-fraction]\" data-toc-modified-id=\"Figure-3B-[Cooperatively-bound-fraction]-2.9\"><span class=\"toc-item-num\">2.9&nbsp;&nbsp;</span>Figure 3B [Cooperatively bound fraction]</a></span></li></ul></li><li><span><a href=\"#False-positive-simulations-(Figure-4)\" data-toc-modified-id=\"False-positive-simulations-(Figure-4)-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>False positive simulations (Figure 4)</a></span><ul class=\"toc-item\"><li><span><a href=\"#Initialization-for-Figure-4\" data-toc-modified-id=\"Initialization-for-Figure-4-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>Initialization for Figure 4</a></span></li><li><span><a href=\"#Figure-4A-[Simulation-with-false-positives-and-sample-ROC]\" data-toc-modified-id=\"Figure-4A-[Simulation-with-false-positives-and-sample-ROC]-3.2\"><span class=\"toc-item-num\">3.2&nbsp;&nbsp;</span>Figure 4A [Simulation with false-positives and sample ROC]</a></span></li><li><span><a href=\"#Figure-4B-[Extraction-heterogeneity-impact-on-auROC]\" data-toc-modified-id=\"Figure-4B-[Extraction-heterogeneity-impact-on-auROC]-3.3\"><span class=\"toc-item-num\">3.3&nbsp;&nbsp;</span>Figure 4B [Extraction heterogeneity impact on auROC]</a></span></li><li><span><a href=\"#Figure-4B-[Amplification-heterogeneity-impact-on-auROC]\" data-toc-modified-id=\"Figure-4B-[Amplification-heterogeneity-impact-on-auROC]-3.4\"><span class=\"toc-item-num\">3.4&nbsp;&nbsp;</span>Figure 4B [Amplification heterogeneity impact on auROC]</a></span></li><li><span><a href=\"#[Figure-4E]-F-P-occupancy-dependence\" data-toc-modified-id=\"[Figure-4E]-F-P-occupancy-dependence-3.5\"><span class=\"toc-item-num\">3.5&nbsp;&nbsp;</span>[Figure 4E] F-P occupancy dependence</a></span></li><li><span><a href=\"#[Figure-4F]-Coverage-dependence\" data-toc-modified-id=\"[Figure-4F]-Coverage-dependence-3.6\"><span class=\"toc-item-num\">3.6&nbsp;&nbsp;</span>[Figure 4F] Coverage dependence</a></span></li></ul></li><li><span><a href=\"#ChIP-seq-Fidelity-dependence-on-experimental-parameters-[Figure-2-&amp;-3]\" data-toc-modified-id=\"ChIP-seq-Fidelity-dependence-on-experimental-parameters-[Figure-2-&amp;-3]-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>ChIP-seq Fidelity dependence on experimental parameters [Figure 2 &amp; 3]</a></span><ul class=\"toc-item\"><li><span><a href=\"#[Figure-2C]-Sample\" data-toc-modified-id=\"[Figure-2C]-Sample-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>[Figure 2C] Sample</a></span></li><li><span><a href=\"#[Figure-2C]-Extraction-efficiency\" data-toc-modified-id=\"[Figure-2C]-Extraction-efficiency-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>[Figure 2C] Extraction efficiency</a></span></li><li><span><a href=\"#[Figure-2C]-PCR-amplification-efficiency\" data-toc-modified-id=\"[Figure-2C]-PCR-amplification-efficiency-4.3\"><span class=\"toc-item-num\">4.3&nbsp;&nbsp;</span>[Figure 2C] PCR amplification efficiency</a></span></li><li><span><a href=\"#[Figure-3C]-Indirect-binding\" data-toc-modified-id=\"[Figure-3C]-Indirect-binding-4.4\"><span class=\"toc-item-num\">4.4&nbsp;&nbsp;</span>[Figure 3C] Indirect binding</a></span></li><li><span><a href=\"#[Figure-3C]-Cooperative-binding\" data-toc-modified-id=\"[Figure-3C]-Cooperative-binding-4.5\"><span class=\"toc-item-num\">4.5&nbsp;&nbsp;</span>[Figure 3C] Cooperative binding</a></span></li></ul></li><li><span><a href=\"#Chromatin-accessibility-[Supplement]\" data-toc-modified-id=\"Chromatin-accessibility-[Supplement]-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Chromatin accessibility [Supplement]</a></span><ul class=\"toc-item\"><li><span><a href=\"#Area-under-ROC\" data-toc-modified-id=\"Area-under-ROC-5.1\"><span class=\"toc-item-num\">5.1&nbsp;&nbsp;</span>Area under ROC</a></span></li></ul></li><li><span><a href=\"#Estimated-number-of-replicates-needed-to-infer-from-ChIP-seq-data-[Figure-5]\" data-toc-modified-id=\"Estimated-number-of-replicates-needed-to-infer-from-ChIP-seq-data-[Figure-5]-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Estimated number of replicates needed to infer from ChIP-seq data [Figure 5]</a></span><ul class=\"toc-item\"><li><span><a href=\"#Plotting-results-for-number-of-replicates-needed-for-binding-energy\" data-toc-modified-id=\"Plotting-results-for-number-of-replicates-needed-for-binding-energy-6.1\"><span class=\"toc-item-num\">6.1&nbsp;&nbsp;</span>Plotting results for number of replicates needed for binding energy</a></span></li><li><span><a href=\"#Computing-distribution-of-energy-differences-between-two-single-site-variants-of-a-binding-site-sequence\" data-toc-modified-id=\"Computing-distribution-of-energy-differences-between-two-single-site-variants-of-a-binding-site-sequence-6.2\"><span class=\"toc-item-num\">6.2&nbsp;&nbsp;</span>Computing distribution of energy differences between two single-site variants of a binding site sequence</a></span></li></ul></li><li><span><a href=\"#FASTA-generation-example\" data-toc-modified-id=\"FASTA-generation-example-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>FASTA generation example</a></span></li><li><span><a href=\"#[Supplement]-Compute-K-L-distance-of-motif-due-to-extraction-and-PCR-efficiency-changes-for-different-TFs\" data-toc-modified-id=\"[Supplement]-Compute-K-L-distance-of-motif-due-to-extraction-and-PCR-efficiency-changes-for-different-TFs-8\"><span class=\"toc-item-num\">8&nbsp;&nbsp;</span>[Supplement] Compute K-L distance of motif due to extraction and PCR efficiency changes for different TFs</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See https://github.com/vishakad/chipulate#getting-started in order to run the code in this Python3 notebook.\n",
    "\n",
    "The module sklearn is needed in addition to the modules listed in the URL above. sklearn is required for computing the area under the ROC curve in Figure 3 of our manuscript."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Initialization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-03T09:36:57.405723Z",
     "start_time": "2019-01-03T09:36:57.289980Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using matplotlib backend: TkAgg\n",
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import matplotlib\n",
    "matplotlib.rcParams['svg.fonttype'] = 'none'\n",
    "from matplotlib import colors as mplcolors\n",
    "import scipy\n",
    "import numpy as np\n",
    "import util\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "from matplotlib import cm\n",
    "import chip_scratch as cs\n",
    "import MotifTable\n",
    "#import sklearn.metrics\n",
    "import plots\n",
    "import distributions\n",
    "%matplotlib\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "def getInputEnergy( occProbs, mfold, E0=0 ):\n",
    "    pAvg = np.mean( occProbs )\n",
    "    pInput = pAvg/(1.0*mfold)\n",
    "    inputEnergy = E0 + np.log( (1/pInput) - 1 )\n",
    "    return inputEnergy\n",
    "\n",
    "def getFPpowerLaw( oldBeta, oldEmax, g ):\n",
    "    m = distributions.truncPowerLawMean( oldBeta, 0, oldEmax )\n",
    "    v = distributions.truncPowerLawVar( oldBeta, 0, oldEmax )\n",
    "    \n",
    "    r = (v/(g*g*m*m) + 1)\n",
    "\n",
    "    beta = 3 + 2/(r-2)\n",
    "    Emax = g * m * ( 2 - beta )\n",
    "    \n",
    "    return [beta,Emax]\n",
    "\n",
    "def getPaperFigureHandle( paperFigStr ):\n",
    "    if plt.fignum_exists( paperFigStr ):\n",
    "        return paperFigures[paperFigStr]\n",
    "    else:\n",
    "        return createFigure( paperFigStr )\n",
    "\n",
    "paperFigures = {}\n",
    "def createFigure( paperFigStr ):\n",
    "    global paperFigures\n",
    "    if paperFigStr == 'Figure 3':\n",
    "        paperFigures[paperFigStr] = plt.subplots( nrows=2,ncols=3, num='Figure 3' )\n",
    "    elif paperFigStr == 'Figure 4':\n",
    "        paperFigures[paperFigStr] = plt.subplots( nrows=2,ncols=3, num='Figure 4' )\n",
    "    elif paperFigStr == 'Figure 5':\n",
    "        paperFigures[paperFigStr] = plt.subplots( nrows=2,ncols=3, num='Figure 5' )        \n",
    "    elif paperFigStr == 'Figure 2':\n",
    "        paperFigures[paperFigStr] = plt.subplots( nrows=2, ncols=3, num='Figure 2' )\n",
    "    elif paperFigStr == 'Figure S2':\n",
    "        paperFigures[paperFigStr] = plt.subplots( nrows=2, ncols=3, num='Figure S2' )\n",
    "    elif paperFigStr == 'Figure MI':\n",
    "        paperFigures[paperFigStr] = plt.subplots( nrows=1, ncols=2, num='Figure MI' )\n",
    "    elif paperFigStr == 'Figure S1':\n",
    "        paperFigures[paperFigStr] = plt.subplots( nrows=4, ncols=4, num='Figure S1', sharey=True )        \n",
    "    elif paperFigStr == 'Figure 1':\n",
    "        paperFigures[paperFigStr] = plt.subplots( nrows=3, ncols=2, num='Figure 1', sharex=True )\n",
    "        \n",
    "    for ax in paperFigures[paperFigStr][1].ravel():\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        \n",
    "    return paperFigures[paperFigStr]\n",
    "\n",
    "yeastBg = [3.093e-01,1.907e-01,1.907e-01,3.093e-01] \n",
    "cbcolors = {'sky blue': (86/255.0,180/255.0,233/255.0)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Change in motif information with ChIP-seq and biological parameters (Figure 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialization for Figure 2\n",
    "\n",
    "Note : The first run of the following code will create all the PCR amplification distributions. Since the number of cycles employed\n",
    "in the manuscript is 15, this may take over an hour to run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "tf = 'Tye7'\n",
    "baselineEnergies, baselineSequences, motifSequences, pfmRef = cs.getBaselineMotif(chemicalPotential=3,numLocations=1000,tf=tf )\n",
    "motif = MotifTable.MotifTable( tf )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Motif information and fidelity content decreases with increasing TF chemical potential [Supplement]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "chemicalPotentials = [1,2,3,4,5,6]\n",
    "numLocations = len( baselineEnergies )\n",
    "\n",
    "numReps = 10\n",
    "motifInformation = np.zeros( (len(chemicalPotentials),numReps) )\n",
    "informationContent = np.zeros( (len(chemicalPotentials),numReps))\n",
    "muIdx = 0\n",
    "\n",
    "quantiles = [[0,0.25],[0.25,0.5],[0.5,0.75],[0.75,1.0],[0,1.0]]\n",
    "ratioColumn = 'ratio'\n",
    "\n",
    "pExtChip = distributions.truncNorm( a=0, b=1, mu=0.5, sigma=0.05, size=np.int(numLocations) )\n",
    "fidelityMat = {}\n",
    "for q in quantiles:\n",
    "    fidelityMat[str(q)] = np.zeros((len(chemicalPotentials),numReps))\n",
    "\n",
    "chemIdx = 0\n",
    "fig, axes = getPaperFigureHandle('Figure 2')\n",
    "\n",
    "ax = axes[0,0]\n",
    "ax.clear()\n",
    "for chemicalPotential in chemicalPotentials:\n",
    "    print( \"{}/{}\".format( chemIdx, len(chemicalPotentials) ) )\n",
    "    for rep in range(numReps):\n",
    "        genome = cs.performChipSeq( baselineSequences, baselineEnergies, pExt=pExtChip, chemicalPotential=chemicalPotential, numCells=100000 ) \n",
    "\n",
    "        if rep == 0 and chemicalPotential in [2,4,6]:\n",
    "            ax.scatter( genome['energy_A'], genome['ratio'], label='$\\mu$ = {}'.format( chemicalPotential ), s=3 )\n",
    "\n",
    "        genome = genome.sort_values(by='ratio',ascending=False)\n",
    "        \n",
    "        learnGenome = genome.head( int(0.1*numLocations ) )\n",
    "\n",
    "        pwmSingle, pcmSingle, pfmSingle = cs.findPWM( learnGenome['sequence'].values )\n",
    "        bgMatrix = np.reshape( np.repeat( yeastBg, pfmSingle.shape[1] ), (4,pfmSingle.shape[1]) )\n",
    "        motifInformation[chemIdx,rep] = MotifTable.klDist( pfmRef, pfmSingle )/pfmRef.shape[1]\n",
    "        informationContent[chemIdx,rep] = np.sum( pfmSingle * np.log2( pfmSingle/bgMatrix ), axis=0 ).sum()\n",
    "        \n",
    "        fidelityValues = util.computeFidelity( genome, ratioColumn=ratioColumn )\n",
    "        qIdx = 0\n",
    "        for q in quantiles:\n",
    "            fidelityMat[str(q)][chemIdx,rep] = fidelityValues[qIdx]\n",
    "            qIdx += 1\n",
    "\n",
    "    chemIdx += 1\n",
    "\n",
    "ax.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fig, axes = getPaperFigureHandle('Figure S2')\n",
    "ax = axes[0,1]\n",
    "ax.clear()\n",
    "plt.sca(ax)\n",
    "\n",
    "means = motifInformation.mean(axis=1)\n",
    "stds = motifInformation.std(axis=1)\n",
    "ax.errorbar( chemicalPotentials, means, yerr=stds, marker='.' )\n",
    "ax.set_xlabel('Chemical potential',fontsize=10)\n",
    "#ax.set_ylabel('Mean K-L distance\\n from baseline (bits)',fontsize=8)\n",
    "\n",
    "ax.tick_params(labelsize=10)\n",
    "ax.yaxis.set_major_locator( mpl.ticker.MaxNLocator(5) )\n",
    "\n",
    "ax = axes[0,2]\n",
    "ax.clear()\n",
    "plt.sca(ax)\n",
    "idx = 0\n",
    "lineStyles = ['dotted','dashed','solid']\n",
    "labels = ['Bottom 25','25-50','50-75','Top 25','All']\n",
    "lIdx = 0\n",
    "for q in quantiles:\n",
    "    if idx in [0,3,4]:\n",
    "        means = fidelityMat[str(q)].mean(axis=1)\n",
    "        stds = fidelityMat[str(q)].std(axis=1)\n",
    "\n",
    "        ax.errorbar( chemicalPotentials, means, yerr=stds, marker='.', color='black', linestyle=lineStyles[lIdx], label=labels[idx] )\n",
    "        ax.set_xlabel('Chemical potential', fontsize=10)\n",
    "        ax.set_ylabel('Fidelity',fontsize=10)\n",
    "        lIdx += 1\n",
    "\n",
    "        ax.set_ylim([0,1])\n",
    "    idx += 1\n",
    "ax.legend()\n",
    "\n",
    "ax = axes[1,0]\n",
    "ax.clear()\n",
    "means = informationContent.mean(axis=1)\n",
    "stds = informationContent.std(axis=1)\n",
    "ax.errorbar( chemicalPotentials, means, yerr=stds, marker='.' )\n",
    "ax.set_xlabel('Chemical potential',fontsize=10)\n",
    "ax.set_ylabel('Motif information content (bits)',fontsize=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 2A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fig, axes = getPaperFigureHandle('Figure 2')\n",
    "ax = axes[0,0]\n",
    "ax.clear()\n",
    "\n",
    "chemicalPotential = 3\n",
    "baselineEnergies, baselineSequences, pfmRef = cs.getBaselineMotif(chemicalPotential=chemicalPotential)\n",
    "tf = 'Tye7'\n",
    "motif = MotifTable.MotifTable( tf )\n",
    "\n",
    "numLocations = 1000\n",
    "pExt = distributions.truncNorm( a=0, b=1, mu=0.5, sigma=0.05, size=numLocations)\n",
    "genome = cs.performChipSeq( spEnergies=baselineEnergies, sequences=baselineSequences, chemicalPotential=chemicalPotential, numCells=100000 )\n",
    "\n",
    "ratioQ = genome['ratio'].quantile(0.9)\n",
    "learnGenome = genome.query('ratio > {}'.format( ratioQ ))\n",
    "ax.scatter( genome.query('ratio < {}'.format( ratioQ ))['energy_A'].values, genome.query('ratio < {}'.format(ratioQ))['ratio'].values, color='gray', marker='.', s=2 )\n",
    "ax.scatter( learnGenome['energy_A'].values, learnGenome['ratio'].values, color='black', marker='.', s=2 )\n",
    "ax.plot( [genome['energy_A'].min(),genome['energy_A'].max()], [ratioQ,ratioQ], color='black', linestyle='--' )\n",
    "ax.yaxis.set_major_locator( mpl.ticker.MaxNLocator(5) )\n",
    "ax.set_xlabel('Binding energy')\n",
    "ax.set_ylabel('Read count ratio')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 2B [PCR amplification]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "variationType = 'powerLaw'\n",
    "\n",
    "if variationType == 'normal':\n",
    "    muList = [10,100,1000,10000]\n",
    "    cvList = [0.0,0.5,1.0]\n",
    "elif variationType == 'powerLaw':\n",
    "    muList = [1.0]\n",
    "    cvList = [-0.90,-0.5,-0.1,0.8,4]\n",
    "    \n",
    "#baselineAmp = 0.01\n",
    "numLocations = len( baselineEnergies )\n",
    "muRange = np.zeros_like( cvList )\n",
    "stdRange = np.zeros_like( cvList )\n",
    "chemicalPotential = 3\n",
    "#vRange = np.zeros_like( cvList )\n",
    "\n",
    "pExt = 1.0\n",
    "\n",
    "numReps = 10\n",
    "motifInformation = np.zeros( (len(muList), len(cvList),numReps) )\n",
    "ampRatioRecords = {}\n",
    "\n",
    "pcrCycles = 15\n",
    "muIdx = 0\n",
    "for muAmp in muList:\n",
    "    widthIdx = 0\n",
    "    for width in cvList:\n",
    "        if variationType == 'normal':\n",
    "            if width >0:\n",
    "                muAmpWidth = width*muAmp\n",
    "                ampRatio = distributions.truncNorm( a=1, b=2**pcrCycles, mu=muAmp, sigma=width*muAmp, size=np.int(numLocations) )\n",
    "            else:\n",
    "                ampRatio = muAmp\n",
    "        elif variationType == 'powerLaw':\n",
    "            pAmp = distributions.truncPowerLawRVs( width, 1, 100, size=numLocations, algorithm='scipy' )/100\n",
    "            ampRatio = np.power( 1 + pAmp, pcrCycles )\n",
    "\n",
    "            muRange[widthIdx] = np.mean( ampRatio )\n",
    "            stdRange[widthIdx] = np.std( ampRatio )\n",
    "            ampRatioRecords[str(width)] = ampRatio\n",
    "\n",
    "        print( np.mean( ampRatio ))\n",
    "        print( \"{}/{},{}/{}\".format( muIdx, len(muList), widthIdx, len(cvList) ) )\n",
    "        for rep in range(numReps):\n",
    "            genome = cs.performChipSeq( baselineSequences, baselineEnergies, ampRatio=ampRatio, pExt=pExt, pcrCycles=pcrCycles, \n",
    "                                       chemicalPotential=chemicalPotential, numCells=100000 )\n",
    "\n",
    "            \n",
    "            genome = genome.sort_values(by='ratio',ascending=False)\n",
    "            learnGenome = genome.head( np.int(0.1*numLocations) )\n",
    "\n",
    "            pwmSingle, pcmSingle, pfmSingle = cs.findPWM( learnGenome['sequence'].values )\n",
    "\n",
    "            motifInformation[muIdx,widthIdx,rep] = MotifTable.klDist( pfmRef, pfmSingle )/pfmRef.shape[1]\n",
    "            \n",
    "        widthIdx += 1\n",
    "    muIdx += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "if variationType == 'normal':\n",
    "    fig, axes = getPaperFigureHandle('Figure 2')\n",
    "elif variationType == 'powerLaw':\n",
    "    fig, axes = getPaperFigureHandle('Figure S2')    \n",
    "    \n",
    "ax = axes[0,1]\n",
    "ax.clear()\n",
    "plt.sca(ax)\n",
    "means = motifInformation.mean(axis=2)\n",
    "stds = motifInformation.std(axis=2)\n",
    "\n",
    "if variationType == 'normal':\n",
    "    _ = plots.manyErrorBarsPlot( muList, means, stds, cvList, cbarText='Coefficient of variation', xIsMean=True,\n",
    "                        xAxisText='Mean amplification ratio', yAxisText='Mean K-L distance from baseline (bits)', cmap='jet', fontsize=10, plotCbar=False )\n",
    "elif variationType == 'powerLaw':\n",
    "    idxes = np.argsort( muRange )\n",
    "    ax.errorbar( muRange[idxes], means[0][idxes], yerr=stds[0], marker='.' )    \n",
    "    ax.set_xlabel('Mean amplification ratio')\n",
    "    ax.set_ylabel('Mean K-L distance\\n from baseline (bits)',fontsize=10)\n",
    "\n",
    "ax.yaxis.set_major_locator( mpl.ticker.MaxNLocator(5) )\n",
    "ax.set_xscale('log')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 2B [Extraction efficiency]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "variationType = 'powerLaw'\n",
    "\n",
    "if variationType == 'normal':\n",
    "    muList = [0.001,0.01,0.1,0.5,0.9]\n",
    "    cvList = [0,0.5,1.0]\n",
    "elif variationType == 'powerLaw':\n",
    "    muList = [1.0]\n",
    "    cvList = [-0.999,-0.9,-0.5,0.90,4]\n",
    "\n",
    "    \n",
    "muRange = np.zeros_like( cvList )\n",
    "chemicalPotential = 3\n",
    "\n",
    "numReps = 10\n",
    "motifInformation = np.zeros( (len(muList), len(cvList),numReps) )\n",
    "muIdx = 0\n",
    "\n",
    "for muExt in muList:\n",
    "    widthIdx = 0\n",
    "    for width in cvList:\n",
    "        if width > 0 and variationType == 'normal':\n",
    "            pExtChip = distributions.truncNorm( a=0, b=1, mu=muExt, sigma=width*muExt, size=np.int(numLocations) )\n",
    "        else:\n",
    "            if variationType == 'normal':\n",
    "                pExtChip = muExt\n",
    "            elif variationType == 'powerLaw':\n",
    "                pExtChip = distributions.truncPowerLawRVs( width, 0, 1, size=np.int(numLocations), algorithm='scipy' )\n",
    "                muRange[widthIdx] = np.mean( pExtChip )\n",
    "        \n",
    "        pExtControl = pExtChip\n",
    "        print( \"{}/{},{}/{}\".format( muIdx, len(muList), widthIdx, len(cvList) ) )\n",
    "        for rep in range(numReps):\n",
    "            genome = cs.performChipSeq( baselineSequences, baselineEnergies, pExt=pExtChip, \n",
    "                                       numCells=100000, chemicalPotential=chemicalPotential ) \n",
    "\n",
    "            genome = genome.sort_values(by='ratio',ascending=False)\n",
    "            learnGenome = genome.head( np.int(0.1*numLocations) )\n",
    "\n",
    "            pwmSingle, pcmSingle, pfmSingle = cs.findPWM( learnGenome['sequence'].values )\n",
    "            motifInformation[muIdx,widthIdx,rep] = MotifTable.klDist( pfmRef, pfmSingle )/pfmRef.shape[1]\n",
    "            \n",
    "        widthIdx += 1\n",
    "    muIdx += 1\n",
    "\n",
    "if variationType == 'normal':\n",
    "    fig, axes = getPaperFigureHandle('Figure 2')\n",
    "elif variationType == 'powerLaw':\n",
    "    fig, axes = getPaperFigureHandle('Figure S2')    \n",
    "    \n",
    "ax = axes[0,2]\n",
    "ax.clear()\n",
    "plt.sca(ax)\n",
    "\n",
    "means = motifInformation.mean(axis=2)\n",
    "stds = motifInformation.std(axis=2)\n",
    "if variationType == 'normal':\n",
    "    _ = plots.manyErrorBarsPlot( muList, means, stds, cvList, cbarText='Coefficient of variation',xIsMean=True,\n",
    "                            xAxisText='Mean extraction efficiency', yAxisText='Mean K-L distance\\n from baseline (bits)', cmap='jet',fontsize=10, plotCbar=False) \n",
    "elif variationType == 'powerLaw':\n",
    "    ax.errorbar( muRange, means[0], yerr=stds[0], marker='.' )\n",
    "    ax.set_xlabel('Mean extraction efficiency',fontsize=10)\n",
    "    ax.set_ylabel('Mean K-L distance\\n from baseline (bits)',fontsize=10)\n",
    "\n",
    "ax.tick_params(labelsize=10)\n",
    "ax.yaxis.set_major_locator( mpl.ticker.MaxNLocator(5) )\n",
    "ax.set_xscale('log')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "if variationType == 'normal':\n",
    "    fig, axes = getPaperFigureHandle('Figure 2')\n",
    "elif variationType == 'powerLaw':\n",
    "    fig, axes = getPaperFigureHandle('Figure S2')    \n",
    "    \n",
    "ax = axes[0,2]\n",
    "ax.clear()\n",
    "plt.sca(ax)\n",
    "\n",
    "means = motifInformation.mean(axis=2)\n",
    "stds = motifInformation.std(axis=2)\n",
    "if variationType == 'normal':\n",
    "    _ = plots.manyErrorBarsPlot( muList, means, stds, cvList, cbarText='Coefficient of variation',xIsMean=True,\n",
    "                            xAxisText='Mean extraction efficiency', yAxisText='Mean K-L distance\\n from baseline (bits)', cmap='jet',fontsize=10, plotCbar=False) \n",
    "elif variationType == 'powerLaw':\n",
    "    ax.errorbar( muRange, means[0], yerr=stds[0], marker='.' )\n",
    "    ax.set_xlabel('Mean extraction efficiency',fontsize=10)\n",
    "    ax.set_ylabel('Mean K-L distance\\n from baseline (bits)',fontsize=10)\n",
    "\n",
    "ax.tick_params(labelsize=10)\n",
    "ax.yaxis.set_major_locator( mpl.ticker.MaxNLocator(5) )\n",
    "ax.set_xscale('log')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 3A [Top, sample output, indirect binding]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "numLocations = 1000\n",
    "\n",
    "pExt = distributions.truncNorm( a=0, b=1.0, mu=0.5, sigma=0.05, size=numLocations )\n",
    "\n",
    "fracIndirect = 0.3\n",
    "mainTf = 'Tye7'\n",
    "\n",
    "indirectIdx = 0\n",
    "\n",
    "directMotif = MotifTable.MotifTable( mainTf )\n",
    "energiesSampled = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numLocations )\n",
    "spEnergies, sequences = directMotif.sampleFromDistribution( energiesSampled )\n",
    "\n",
    "secondTFspEnergies = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numLocations )\n",
    "backgroundDf = cs.makeBackgroundSequences( numLocations, mainTf )\n",
    "\n",
    "indirectLocations = np.arange(int(fracIndirect*numLocations))\n",
    "chemicalPotential = 3\n",
    "secondTFchemicalPotential = 3\n",
    "genome = cs.performChipSeq( spEnergies=spEnergies, secondTFspEnergies=secondTFspEnergies, indirectLocations=indirectLocations,\n",
    "                          chemicalPotential=chemicalPotential, secondTFchemicalPotential=secondTFchemicalPotential )\n",
    "\n",
    "fig, axes = getPaperFigureHandle('Figure 3')\n",
    "ax = axes[0,0]\n",
    "ax.clear()\n",
    "\n",
    "genome.loc[:,'ratio'] = genome.eval('unique_chip_reads/unique_control_reads')\n",
    "directDf = genome.query('binding == \"direct\"')\n",
    "indirectDf = genome.query('binding == \"indirect\"')\n",
    "ax.scatter( directDf['energy_A'], directDf['ratio'].values, marker='.', label='Direct', s=3 )\n",
    "ax.scatter( indirectDf['energy_A'], indirectDf['ratio'].values, marker='.', label='Indirect', s=3 )\n",
    "ax.set_xlabel('Binding energy')\n",
    "ax.set_ylabel('Read count ratio')\n",
    "ax.legend() \n",
    "\n",
    "ax.yaxis.set_major_locator( mpl.ticker.MaxNLocator(5) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 3A [Bottom, sample output, cooperative binding]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "numLocations = 1000\n",
    "\n",
    "pExt = distributions.truncNorm( a=0, b=1.0, mu=0.5, sigma=0.05, size=numLocations )\n",
    "\n",
    "numCoop = 300\n",
    "mainTf = 'Tye7'\n",
    "coopTf = 'Cbf1'\n",
    "indirectIdx = 0\n",
    "\n",
    "directMotif = MotifTable.MotifTable( mainTf )\n",
    "energiesSampled = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numLocations )\n",
    "spEnergies, sequences = directMotif.sampleFromDistribution( energiesSampled )\n",
    "\n",
    "coopDf, coopMotifs = cs.makeCoopSequences( numLocations, [mainTf,coopTf], [['powerLaw',0.5,0,10],['powerLaw',0.5,0,10]] ) #This will choose a single random indirectly bound partner\n",
    "secondSequences = coopDf['sequence_2'].values\n",
    "secondTFspEnergies = coopDf['sp_energy_2'].values\n",
    "\n",
    "tfIntEnergies = np.append( np.ones( numCoop )*(-6), np.zeros( numLocations - numCoop ))\n",
    "\n",
    "chemicalPotential = 3\n",
    "secondTFchemicalPotential = 3\n",
    "\n",
    "genome = cs.performChipSeq( spEnergies=spEnergies, secondTFspEnergies=secondTFspEnergies,  secondTFintEnergies=tfIntEnergies,\n",
    "                          chemicalPotential=chemicalPotential, secondTFchemicalPotential=secondTFchemicalPotential, numCells=10000 )\n",
    "\n",
    "fig, axes = getPaperFigureHandle('Figure 3')\n",
    "ax = axes[1,0]\n",
    "ax.clear()\n",
    "\n",
    "genome.loc[:,'ratio'] = genome.eval('unique_chip_reads/unique_control_reads')\n",
    "q = genome['ratio'].quantile( 0.9 )\n",
    "directDf = genome.query('binding == \"direct\"')\n",
    "coopDf = genome.query('binding == \"cooperative\"')\n",
    "ax.scatter( directDf['energy_A'], directDf['ratio'].values, marker='.', label='Independent', s=3 )\n",
    "ax.scatter( coopDf['energy_A'], coopDf['ratio'].values, marker='.', label='Cooperative', s=3 )\n",
    "ax.set_xlabel('Binding energy')\n",
    "ax.set_ylabel('Read count ratio')\n",
    "ax.legend() \n",
    "\n",
    "ax.yaxis.set_major_locator( mpl.ticker.MaxNLocator(5) )\n",
    "print( scipy.stats.pearsonr( coopDf['energy_A'].values, coopDf['ratio'].values ) )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 3B [Middle Row, Left, Indirect binding]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "numLocations = 1000\n",
    "\n",
    "pExt = distributions.truncNorm( a=0, b=1.0, mu=0.5, sigma=0.05, size=numLocations )\n",
    "numReps = 10\n",
    "\n",
    "fracIndirectSet = [0.001,0.01,0.1,0.2,0.3,0.4,0.5]\n",
    "mainTf = 'Tye7'\n",
    "\n",
    "indirectIdx = 0\n",
    "\n",
    "directMotif = MotifTable.MotifTable( mainTf )\n",
    "energiesSampled = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numLocations )\n",
    "spEnergies, sequences = directMotif.sampleFromDistribution( energiesSampled )\n",
    "\n",
    "genome = cs.performChipSeq( sequences, spEnergies, pExt=pExt, numCells=100000 )\n",
    "\n",
    "learnGenome = genome.sort_values(by='ratio',ascending=False).head(np.int(0.1*numLocations))\n",
    "pwmRef, pcmRef, pfmRef = cs.findPWM( learnGenome['sequence'].values )\n",
    "\n",
    "energyRange = [0,10]\n",
    "lineStyles = ['solid']\n",
    "colors = ['black']\n",
    "fig, axes = getPaperFigureHandle( 'Figure 3' )\n",
    "ax = axes[0,1]\n",
    "ax.clear()\n",
    "\n",
    "secondTFspEnergies = distributions.truncPowerLawRVs( 0.5, energyRange[0], energyRange[1], size=numLocations )\n",
    "backgroundDf = cs.makeBackgroundSequences( numLocations, mainTf )\n",
    "backgroundSequences = backgroundDf['sequence'].values\n",
    "motifInformation = np.zeros( (len( fracIndirectSet ), numReps ) )\n",
    "\n",
    "chemicalPotential = 3\n",
    "secondTFchemicalPotential = 3\n",
    "\n",
    "indirectIdx = 0\n",
    "for fracIndirect in fracIndirectSet:\n",
    "    numIndirect = np.int( fracIndirect*numLocations )\n",
    "\n",
    "    indirectLocations = np.arange( numIndirect )\n",
    "\n",
    "    sequencesToPass = np.copy( sequences )\n",
    "    sequencesToPass[:numIndirect] = backgroundDf['sequence'].values[:numIndirect]\n",
    "\n",
    "    for rep in range(numReps):\n",
    "        print( \"{}/{},{}/{} : f = {}\".format( indirectIdx+1, len(fracIndirectSet), rep+1, numReps, fracIndirect))\n",
    "        genome = cs.performChipSeq( sequences, spEnergies, numCells=100000, secondTFspEnergies=secondTFspEnergies, \n",
    "                                   indirectLocations=indirectLocations, chemicalPotential=chemicalPotential,  \n",
    "                                  secondTFchemicalPotential=secondTFchemicalPotential )\n",
    "\n",
    "        learnGenome = genome.sort_values(by='ratio',ascending=False).head(np.int(0.1*numLocations))\n",
    "        pwmDouble, pcmDouble, pfmDouble = cs.findPWM( learnGenome['sequence'].values )\n",
    "        motifInformation[indirectIdx,rep] = MotifTable.klDist( pfmRef, pfmDouble )/pfmRef.shape[1]\n",
    "    indirectIdx += 1\n",
    "\n",
    "ax.errorbar( fracIndirectSet, motifInformation.mean(axis=1), marker='.', color='black', ls=lineStyles[0], yerr=motifInformation.std(axis=1) )\n",
    "ax.set_xlabel('Fraction indirectly bound',fontsize=10)\n",
    "#ax.set_ylabel('Mean K-L distance from single motif (bits)',fontsize=10)\n",
    "ax.yaxis.set_major_locator( mpl.ticker.MaxNLocator(5) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "lineStyles = ['solid']\n",
    "colors = ['black']\n",
    "fig, axes = getPaperFigureHandle( 'Figure 3' )\n",
    "ax = axes[0,1]\n",
    "ax.clear()\n",
    "\n",
    "ax.errorbar( fracIndirectSet, motifInformation.mean(axis=1), marker='.', color='black', ls=lineStyles[0], yerr=motifInformation.std(axis=1) )\n",
    "ax.set_xlabel('Fraction indirectly bound',fontsize=10)\n",
    "#ax.set_ylabel('Mean K-L distance from single motif (bits)',fontsize=10)\n",
    "ax.yaxis.set_major_locator( mpl.ticker.MaxNLocator(5) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 3B [Cooperatively bound fraction]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "numLocations = 1000\n",
    "\n",
    "numReps = 10\n",
    "#fracCoopSet = np.arange( 0.001, 1.0, 0.1 )\n",
    "fracCoopSet = [0.001,0.01,0.1,0.2,0.3,0.4,0.5]\n",
    "motifInformation = np.zeros( (len( fracCoopSet ), numReps ) )\n",
    "\n",
    "coopTf = 'Cbf1'\n",
    "mainTf = 'Tye7'\n",
    "intEnergyList = [-6,-2]\n",
    "\n",
    "directDf, directMotif = cs.makeDirectSequences( numLocations, mainTf, ['powerLaw',0.5,0,10] )\n",
    "sequences = directDf['sequence'].values\n",
    "spEnergies = directDf['sp_energy_1'].values\n",
    "pExt = distributions.truncNorm( a=0, b=1, mu=0.5, sigma=0.05, size=numLocations )\n",
    "genome = cs.performChipSeq( sequences, spEnergies, pExt=pExt )\n",
    "\n",
    "learnGenome = genome.sort_values(by='ratio',ascending=False).head(np.int(0.1*numLocations))\n",
    "\n",
    "pwmSingle, pcmSingle, pfmRef = cs.findPWM( learnGenome['sequence'].values )\n",
    "\n",
    "energyRange = [0,10]\n",
    "\n",
    "intEnergyColors = ['black',cbcolors['sky blue']]\n",
    "fig, axes = getPaperFigureHandle( 'Figure 3' )\n",
    "ax = axes[0,2]\n",
    "\n",
    "chemicalPotential = 3\n",
    "secondTFchemicalPotential = 3\n",
    "\n",
    "lineStyles = ['solid']\n",
    "ax.clear()\n",
    "coopDf, coopMotifs = cs.makeCoopSequences( numLocations, [mainTf,coopTf], [['powerLaw',0.5,0,10],['powerLaw',0.5,energyRange[0],energyRange[1]]] ) #This will choose a single random indirectly bound partner\n",
    "secondSequences = coopDf['sequence_2'].values\n",
    "secondTFspEnergies = coopDf['sp_energy_2'].values\n",
    "\n",
    "intEnergyItr = 0\n",
    "for intEnergy in intEnergyList:\n",
    "    motifInformation = np.zeros( (len( fracCoopSet ), numReps ) )\n",
    "    coopIdx = 0\n",
    "    for fracCoop in fracCoopSet:\n",
    "        numCoop = np.int( fracCoop*numLocations )\n",
    "\n",
    "        coopLocations = np.arange( numCoop )\n",
    "        tfIntEnergies = np.append( np.ones( numCoop )*intEnergy, np.zeros( numLocations - numCoop ))\n",
    "\n",
    "        for rep in range(numReps):\n",
    "            print( \"{}/{},{}/{},{}/{}\".format( coopIdx+1, len(fracCoopSet), intEnergyItr, len(intEnergyList), rep+1, numReps))\n",
    "\n",
    "            genome = cs.performChipSeq( sequences, spEnergies, pExt=pExt,secondTFspEnergies=secondTFspEnergies, \n",
    "                                       secondTFintEnergies=tfIntEnergies, chemicalPotential=chemicalPotential, \n",
    "                                       secondTFchemicalPotential=secondTFchemicalPotential, numCells=100000 )\n",
    "\n",
    "            learnGenome = genome.sort_values(by='ratio',ascending=False).head(np.int(0.1*numLocations))\n",
    "\n",
    "            pwmDouble, pcmDouble, pfmDouble = cs.findPWM( learnGenome['sequence'].values )\n",
    "            motifInformation[coopIdx,rep] = MotifTable.klDist( pfmRef, pfmDouble )/pfmRef.shape[1]\n",
    "        coopIdx += 1\n",
    "\n",
    "    ax.errorbar( fracCoopSet, motifInformation.mean(axis=1), color=intEnergyColors[intEnergyItr], ls=lineStyles[0], yerr=motifInformation.std(axis=1), \n",
    "                capthick=2, marker='.', ms=2, label='$\\Delta = {}k_BT$'.format(intEnergy) )\n",
    "    ax.yaxis.set_major_locator( mpl.ticker.MaxNLocator(5) )\n",
    "    ax.set_xlabel('Fraction cooperatively bound',fontsize=10)\n",
    "    axes[0,1].set_ylabel('Mean K-L distance\\nfrom baseline (bits)')\n",
    "    \n",
    "    intEnergyItr += 1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# False positive simulations (Figure 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialization for Figure 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "numLocations = 1000\n",
    "numFPlocations = 1000\n",
    "numLocations += numFPlocations\n",
    "\n",
    "energiesSampled = np.zeros(numLocations)\n",
    "\n",
    "energyThreshold = 1.59 #E0, the binding energy of the unbound state\n",
    "mFold = 10    #This is the occupancy difference between true positive binding sites and \n",
    "\n",
    "beta = 0.5\n",
    "Emax = 6\n",
    "energiesSampled[:numLocations] = distributions.truncPowerLawRVs( beta, 0, Emax, size=numLocations )\n",
    "occProbs = 1/ (1 + np.exp( energiesSampled ) )\n",
    "bgEnergy = getInputEnergy( occProbs, mFold, E0=energyThreshold )\n",
    "\n",
    "fpMfold = 1.42\n",
    "fpBeta, fpEmax = getFPpowerLaw( beta, Emax,  fpMfold )\n",
    "fpEnergies = distributions.truncPowerLawRVs( fpBeta, 0, fpEmax, size=numFPlocations )\n",
    "\n",
    "energiesSampled[(numLocations-numFPlocations):] = fpEnergies\n",
    "pExt = distributions.truncNorm( a=0,b=1,mu=0.5,sigma=0.05, size=numLocations)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 4A [Simulation with false-positives and sample ROC]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "spEnergies = energiesSampled\n",
    "fig, axes = getPaperFigureHandle( 'Figure 4' )\n",
    "ax = axes[0,1]\n",
    "ax.clear()\n",
    "ax.set_xlabel('False positive rate')\n",
    "ax.set_ylabel('True positive rate')\n",
    "\n",
    "chemicalPotential = 3\n",
    "numReps = 1\n",
    "genome = cs.performChipSeq( spEnergies=spEnergies, numFPlocations=numFPlocations, bgEnergy=bgEnergy, pExt=pExt, \n",
    "                           chemicalPotential=chemicalPotential, numCells=100000 )\n",
    "\n",
    "print( 'Mean occupancy ratio between foreground and false positives : {}'.format( genome.query('binding == \"direct\"')['p_occ_chip'].mean()/genome.query('binding == \"false-positive\"')['p_occ_chip'].mean() ) )\n",
    "\n",
    "labels = np.zeros(genome.shape[0],dtype=np.bool)\n",
    "direct = genome['binding'] == \"direct\"\n",
    "labels[direct] = True\n",
    "labels[~direct] = False\n",
    "auROC = sklearn.metrics.roc_auc_score( labels, genome['ratio'].values )\n",
    "fpr, tpr, thresholds = sklearn.metrics.roc_curve( labels, genome['ratio'].values )\n",
    "ax.plot(fpr, tpr, label='Area under ROC = {:.2f}'.format( auROC ) )\n",
    "ax.legend()\n",
    "\n",
    "ax = axes[0,0]\n",
    "ax.clear()\n",
    "ax.set_xlabel('Binding energy')\n",
    "ax.set_ylabel('Read count ratio')\n",
    "\n",
    "tpDf = genome.query('binding == \"direct\"')\n",
    "fpDf = genome.query('binding == \"false-positive\"')\n",
    "ax.scatter( tpDf['energy_A'].values, tpDf['ratio'].values, marker='.', s=3, label='True binding sites', alpha=0.75)\n",
    "ax.scatter( fpDf['energy_A'].values, fpDf['ratio'].values, marker='.', s=3, label='False positives', alpha=0.75 )\n",
    "ax.legend(fontsize=8)           "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 4B [Extraction heterogeneity impact on auROC]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "spEnergies = energiesSampled\n",
    "fig, axes = getPaperFigureHandle( 'Figure 4' )\n",
    "ax = axes[0,2]\n",
    "\n",
    "ax.clear()\n",
    "plt.sca( ax )\n",
    "\n",
    "muList = [0.001,0.01,0.1,0.5,0.9]\n",
    "cvList = [0,0.5,1.0]\n",
    "\n",
    "energiesSampled = distributions.truncPowerLawRVs( beta, 0, Emax, size=numLocations )\n",
    "fpMfoldList = [1.42,2.0]\n",
    "lineStyleList = ['solid','dashed']\n",
    "\n",
    "numReps = 10\n",
    "chemicalPotential = 3\n",
    "fpItr = 0\n",
    "for fpMfold in fpMfoldList:\n",
    "    fpBeta, fpEmax = getFPpowerLaw( beta, Emax,  fpMfold )\n",
    "    fpEnergies = distributions.truncPowerLawRVs( fpBeta, 0, fpEmax, size=numFPlocations )\n",
    "    energiesSampled[(numLocations-numFPlocations):] = fpEnergies\n",
    "    spEnergies = energiesSampled\n",
    "    \n",
    "    occProbs = 1/ (1 + np.exp( energiesSampled - energyThreshold - chemicalPotential ) )\n",
    "    bgEnergy = getInputEnergy( occProbs, mFold, E0=energyThreshold )\n",
    "    \n",
    "    auROCmat = np.zeros( (len(muList), len(cvList),numReps) )\n",
    "        \n",
    "    muIdx = 0\n",
    "    for muExt in muList:\n",
    "        widthIdx = 0\n",
    "        for width in cvList:\n",
    "            if width > 0:\n",
    "                pExtChip = distributions.truncNorm( a=0, b=1, mu=muExt, sigma=width*muExt, size=np.int(numLocations) )\n",
    "                #pExtChip = powerLaw( width, 0, 1, size=np.int(numLocations) )            \n",
    "            else:\n",
    "                pExtChip = muExt\n",
    "\n",
    "            pExtControl = pExtChip\n",
    "            print( \"{}/{},{}/{}\".format( muIdx, len(muList), widthIdx, len(cvList) ) )\n",
    "            for rep in range(numReps):\n",
    "\n",
    "                genome = cs.performChipSeq( spEnergies=spEnergies, pExt=pExtChip, bgEnergy=bgEnergy, numFPlocations=numFPlocations, \n",
    "                                           chemicalPotential=chemicalPotential ) \n",
    "                nanInfs = np.isnan( genome['ratio'].values ) | np.isinf( genome['ratio'].values )\n",
    "                genome = genome.loc[~nanInfs,:]\n",
    "                genome = genome.query('ratio > 0')\n",
    "\n",
    "\n",
    "                auROCmat[muIdx,widthIdx,rep] = util.findAuROC( genome )\n",
    "            widthIdx += 1\n",
    "        muIdx += 1\n",
    "    \n",
    "    means = auROCmat.mean(axis=2)\n",
    "    stds = auROCmat.std(axis=2)\n",
    "    _ = plots.manyErrorBarsPlot( muList, means, stds, cvList, cbarText='Coefficient of variation',xIsMean=True, ls=lineStyleList[fpItr],\n",
    "                            xAxisText='Mean extraction efficiency', yAxisText='Area under ROC', cmap='jet',fontsize=10, plotCbar=False )    \n",
    "    fpItr += 1\n",
    "ax.set_xscale('log')\n",
    "ax.set_ylim(0,1)\n",
    "#ax.grid('on',ls='dashed')\n",
    "#ax.set_axisbelow(True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 4B [Amplification heterogeneity impact on auROC]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "spEnergies = energiesSampled\n",
    "fig, axes = getPaperFigureHandle( 'Figure 4' )\n",
    "ax = axes[1,0]\n",
    "ax.clear()\n",
    "plt.sca( ax )\n",
    "\n",
    "energiesSampled = distributions.truncPowerLawRVs( beta, 0, Emax, size=numLocations )\n",
    "fpMfoldList = [1.42,2.0]\n",
    "lineStyleList = ['solid','dashed']\n",
    "\n",
    "fig, axes = getPaperFigureHandle( 'Figure 4' )\n",
    "ax = axes[1,0]\n",
    "ax.clear()\n",
    "plt.sca( ax )\n",
    "\n",
    "chemicalPotential = 3\n",
    "\n",
    "muList = [10,100,1000,10000]\n",
    "cvList = [0.0,0.5,1.0]\n",
    "\n",
    "numReps = 10\n",
    "pcrCycles = 15\n",
    "fpItr = 0\n",
    "pExtChip = distributions.truncNorm( a=0, b=1, mu=0.5, sigma=0.05, size=numLocations )\n",
    "for fpMfold in fpMfoldList:\n",
    "    fpBeta, fpEmax = getFPpowerLaw( beta, Emax,  fpMfold )\n",
    "    fpEnergies = distributions.truncPowerLawRVs( fpBeta, 0, fpEmax, size=numFPlocations )\n",
    "    energiesSampled[(numLocations-numFPlocations):] = fpEnergies\n",
    "    spEnergies = energiesSampled    \n",
    "    auROCmat = np.zeros( (len(muList), len(cvList),numReps) )\n",
    "    muIdx = 0\n",
    "    \n",
    "    for muAmp in muList:\n",
    "        widthIdx = 0\n",
    "        for width in cvList:\n",
    "            if width > 0:\n",
    "                muAmpWidth = width*muAmp\n",
    "                ampRatio = distributions.truncNorm( a=1, b=2**pcrCycles, mu=muAmp, sigma=width*muAmp, size=np.int(numLocations) )\n",
    "            else:\n",
    "                ampRatio = muAmp\n",
    "\n",
    "            pExtControl = pExtChip\n",
    "            print( \"{}/{},{}/{}\".format( muIdx, len(muList), widthIdx, len(cvList) ) )\n",
    "            for rep in range(numReps):\n",
    "                genome = cs.performChipSeq( spEnergies=spEnergies, pExt=pExtChip, numFPlocations=numFPlocations, numCells=100000, \n",
    "                                           ampRatio=ampRatio, pcrCycles=pcrCycles, bgEnergy=bgEnergy, chemicalPotential=chemicalPotential ) \n",
    "\n",
    "                nanInfs = np.isnan( genome['ratio'].values ) | np.isinf( genome['ratio'].values )\n",
    "                genome = genome.loc[~nanInfs,:]\n",
    "                genome = genome.query('ratio > 0')\n",
    "                \n",
    "                auROCmat[muIdx,widthIdx,rep] = util.findAuROC( genome )\n",
    "            widthIdx += 1\n",
    "        muIdx += 1\n",
    "\n",
    "    means = auROCmat.mean(axis=2)\n",
    "    stds = auROCmat.std(axis=2)\n",
    "    _ = plots.manyErrorBarsPlot( muList, means, stds, cvList, cbarText='Coefficient of variation',xIsMean=True, ls=lineStyleList[fpItr],\n",
    "                                xAxisText='Mean amplification efficiency', yAxisText='Area under ROC', cmap='jet',fontsize=10, plotCbar=False )    \n",
    "    \n",
    "    fpItr += 1\n",
    "ax.set_xscale('log')\n",
    "ax.set_ylim(0,1)\n",
    "#ax.grid('on',ls='dashed')\n",
    "#ax.set_axisbelow(True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## [Figure 4E] F-P occupancy dependence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "numLocations = 1000\n",
    "numFPlocations = 1000\n",
    "numLocations += numFPlocations\n",
    "pExt = distributions.truncNorm( a=0,b=1,mu=0.5,sigma=0.05, size=numLocations)\n",
    "\n",
    "energiesSampled = distributions.truncPowerLawRVs( beta, 0, Emax, size=numLocations )\n",
    "fig, axes = getPaperFigureHandle( 'Figure 4' )\n",
    "ax = axes[1,1]\n",
    "ax.clear()\n",
    "plt.sca( ax )\n",
    "chemicalPotential = 3\n",
    "\n",
    "fpMfoldList = [1,1.2,1.4,1.6,1.8,2,2.5,3,5]\n",
    "numReps = 10\n",
    "pcrCycles = 15\n",
    "auROCmat = np.zeros( (len(fpMfoldList),numReps) )\n",
    "meanOccRatio = np.zeros( len(fpMfoldList ) )\n",
    "mIdx = 0\n",
    "for fpMfold in fpMfoldList:\n",
    "    pExtControl = pExtChip\n",
    "    print( \"{}/{}\".format( mIdx, len(fpMfoldList) ) )\n",
    "    fpBeta, fpEmax = getFPpowerLaw( beta, Emax,  fpMfold )\n",
    "    fpEnergies = distributions.truncPowerLawRVs( fpBeta, 0, fpEmax, size=numFPlocations )\n",
    "\n",
    "    energiesSampled[(numLocations-numFPlocations):] = fpEnergies\n",
    "    for rep in range(numReps):\n",
    "        genome = cs.performChipSeq( spEnergies=energiesSampled, pExt=pExt, numFPlocations=numFPlocations, chemicalPotential=chemicalPotential ) \n",
    "        auROCmat[mIdx,rep] = util.findAuROC( genome )\n",
    "        meanOccRatio[mIdx] = genome.query('binding == \"direct\"')['p_occ_chip'].mean()/genome.query('binding == \"false-positive\"')['p_occ_chip'].mean()\n",
    "\n",
    "    mIdx += 1\n",
    "\n",
    "ax.errorbar( meanOccRatio, auROCmat.mean(axis=1), yerr=auROCmat.std(axis=1),  marker='.', ms=3 )\n",
    "ax.set_xscale('log')\n",
    "ax.set_ylim(0,1)\n",
    "ax.grid('on',ls='dashed')\n",
    "ax.set_axisbelow(True)\n",
    "ax.set_xlabel('Mean occupancy ratio')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [Figure 4F] Coverage dependence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "numLocations = 1000\n",
    "numFPlocations = 1000\n",
    "numLocations += numFPlocations\n",
    "pExt = distributions.truncNorm( a=0,b=1,mu=0.5,sigma=0.05, size=numLocations)\n",
    "\n",
    "fig, axes = getPaperFigureHandle( 'Figure 4' )\n",
    "ax = axes[1,2]\n",
    "ax.clear()\n",
    "plt.sca( ax )\n",
    "\n",
    "chemicalPotential = 3\n",
    "\n",
    "numReps = 10\n",
    "pcrCycles = 15\n",
    "depthList = [10,50,100,500,1000]\n",
    "fpMfoldList = [1.42,2.0]\n",
    "lineStyleList = ['solid','dashed']\n",
    "\n",
    "fpItr = 0\n",
    "for fpMfold in fpMfoldList:\n",
    "    fpBeta, fpEmax = getFPpowerLaw( beta, Emax,  fpMfold )\n",
    "    fpEnergies = distributions.truncPowerLawRVs( fpBeta, 0, fpEmax, size=numFPlocations )\n",
    "    energiesSampled[(numLocations-numFPlocations):] = fpEnergies\n",
    "    spEnergies = energiesSampled\n",
    "    \n",
    "    auROCmat = np.zeros( (len(depthList),numReps) )\n",
    "    mIdx = 0\n",
    "    for depth in depthList:\n",
    "        print( \"{}/{}\".format( mIdx+1, len(depthList) ) )\n",
    "        \n",
    "        for rep in range(numReps):\n",
    "            genome = cs.performChipSeq( spEnergies=energiesSampled, depth=depth, pExt=pExtChip, bgEnergy=bgEnergy, numFPlocations=numFPlocations, chemicalPotential=chemicalPotential ) \n",
    "            \n",
    "            nanInfs = np.isnan( genome['ratio'].values ) | np.isinf( genome['ratio'].values )\n",
    "            genome = genome.loc[~nanInfs,:]\n",
    "            genome = genome.query('ratio > 0')\n",
    "            \n",
    "            auROCmat[mIdx,rep] = util.findAuROC( genome )\n",
    "\n",
    "        mIdx += 1\n",
    "\n",
    "    ax.errorbar( depthList, auROCmat.mean(axis=1), yerr=auROCmat.std(axis=1), marker='.', ms=2, ls=lineStyleList[fpItr], color='black' )\n",
    "    fpItr += 1\n",
    "\n",
    "ax.set_xscale('log')\n",
    "ax.set_ylim(0,1)\n",
    "#ax.grid('on',ls='dashed')\n",
    "#ax.set_axisbelow(True)\n",
    "ax.set_xlabel('Sequencing depth')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ChIP-seq Fidelity dependence on experimental parameters [Figure 2 & 3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [Figure 2C] Sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "numLocations = 1000\n",
    "energiesSampled = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numLocations )\n",
    "\n",
    "spEnergies = energiesSampled\n",
    "\n",
    "fig, axes = getPaperFigureHandle( 'Figure 2' )\n",
    "ax = axes[1,0]\n",
    "ax.clear()\n",
    "\n",
    "chemicalPotential = 3\n",
    "\n",
    "numReps = 10\n",
    "quantiles = [[0,0.25],[0.25,0.5],[0.5,0.75],[0.75,1.0],[0,1.0]]\n",
    "fidelityMat = {}\n",
    "for q in quantiles:\n",
    "    fidelityMat[str(q)] = np.zeros(numReps)\n",
    "\n",
    "labels = ['0-25','25-50','50-75','75-100','All']\n",
    "\n",
    "pExt = distributions.truncNorm( a=0, b=1, mu=0.5, sigma=0.05, size=numLocations)\n",
    "for rep in range(numReps):\n",
    "    print('{}/{}'.format(rep,numReps))\n",
    "    genome = cs.performChipSeq( spEnergies=spEnergies, pExt=pExt, numCells=100000, chemicalPotential=chemicalPotential )\n",
    "    print(\"Finished simulation\")\n",
    "\n",
    "    genome.loc[:,'ratio'] = genome.eval('unique_chip_reads/unique_control_reads').values\n",
    "        \n",
    "    fidelityValues = util.computeFidelity( genome, quantiles=quantiles )        \n",
    "    print( fidelityValues )\n",
    "    qIdx = 0\n",
    "    for q in quantiles:\n",
    "        if rep == 0 and qIdx < 4:\n",
    "            qLow, qHigh = genome['ratio'].quantile(q)\n",
    "            tempDf = genome.query('ratio > {} & ratio < {}'.format( qLow, qHigh ))\n",
    "            ax.scatter( tempDf['ratio'].values, tempDf['energy_A'].values, marker='.', s=3, label=labels[qIdx] )   \n",
    "        fidelityMat[str(q)][rep] = fidelityValues[qIdx]\n",
    "        qIdx += 1\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [Figure 2C] Extraction efficiency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [],
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "variationType = 'powerLaw'\n",
    "\n",
    "if variationType == 'normal':\n",
    "    muExtList = [0.001,0.01,0.1,0.5,0.9]\n",
    "    #sigmaRange = [0.01,0.05,0.25,1.0,2,10]    \n",
    "    sigmaRange = [0,0.5,1.0]\n",
    "elif variationType == 'powerLaw':\n",
    "    muExtList = [1.0]\n",
    "    sigmaRange = [-0.999,-0.9,-0.5,0.90,4]\n",
    "    #sigmaRange = [0.1,0.5,0.9]\n",
    "    #sigmaRange = [-0.5]\n",
    "    \n",
    "    muRange = np.zeros_like( sigmaRange )\n",
    "ratioColumn = 'ratio'\n",
    "\n",
    "numReps = 10\n",
    "numLocations = 1000\n",
    "\n",
    "directMotif = MotifTable.MotifTable( 'Tye7' )\n",
    "energiesSampled = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numLocations )\n",
    "chemicalPotential = 3\n",
    "\n",
    "quantiles = [[0,0.25],[0.25,0.5],[0.5,0.75],[0.75,1.0],[0,1.0]]\n",
    "labels = ['Top 25th percentile','25th - 50th percentile','50th - 75th percentile','75th - 100th percentile','All regions']\n",
    "\n",
    "fidelityMat = {}\n",
    "for q in quantiles:\n",
    "    fidelityMat[str(q)] = np.zeros((len(muExtList),len(sigmaRange),numReps))\n",
    "\n",
    "muIdx = 0\n",
    "figtemp, axestemp = plt.subplots( nrows=1, ncols=6, sharex=True, sharey=True )\n",
    "for mu in muExtList:\n",
    "    sigmaIdx = 0\n",
    "    for sigma in sigmaRange:\n",
    "        if variationType == 'normal':\n",
    "            if sigma > 0:\n",
    "                pExt = distributions.truncNorm( a=0.0, b=1.0, mu=mu, sigma=sigma*mu, size=numLocations )\n",
    "            else:\n",
    "                pExt = mu\n",
    "        elif variationType == 'powerLaw':\n",
    "            pExt = distributions.truncPowerLawRVs( sigma, 0, 1, size=np.int(numLocations), algorithm='scipy' )\n",
    "            muRange[sigmaIdx] = np.std( pExt )/np.mean( pExt )\n",
    "            \n",
    "        repIdx = 0\n",
    "        for rep in range(numReps):\n",
    "            print('{},{},{}'.format(muIdx,sigmaIdx,rep))\n",
    "            if ratioColumn == 'score':\n",
    "                spEnergies, sequences = directMotif.sampleFromDistribution( energiesSampled )                \n",
    "                genome = cs.performChipSeq( spEnergies=spEnergies, pExt=pExt, numCells=100000, chemicalPotential=chemicalPotential )\n",
    "            else:\n",
    "                genome = cs.performChipSeq( spEnergies=energiesSampled, pExt=pExt, numCells=100000, chemicalPotential=chemicalPotential )                \n",
    "\n",
    "            if ratioColumn == 'score':\n",
    "                genome.loc[:,'ratio'] = genome.eval('unique_chip_reads/unique_control_reads').values\n",
    "\n",
    "                learnGenome = genome.sort_values(by='ratio',ascending=False).head(np.int(0.1*numLocations))\n",
    "\n",
    "                pwmDouble, pcmDouble, pfmDouble = cs.findPWM( learnGenome['sequence'].values )\n",
    "                calculatedMotif = MotifTable.MotifTable( 'Tye7', 'yeast', pwmDouble, matType='frequencies')\n",
    "                scores = calculatedMotif.getInformation( genome['sequence'].values )\n",
    "                genome.loc[:,'score'] = scores\n",
    "            \n",
    "            qIdx = 0\n",
    "            fidelityValues = util.computeFidelity( genome, ratioColumn=ratioColumn, minPairs=1000 )\n",
    "            for q in quantiles[:-1]:\n",
    "                fidelityMat[str(q)][muIdx,sigmaIdx,rep] = fidelityValues[qIdx]\n",
    "                qIdx += 1\n",
    "               \n",
    "        sigmaIdx += 1\n",
    "    muIdx += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "idx = 0\n",
    "if variationType == 'normal':\n",
    "    fig, axes = getPaperFigureHandle('Figure 2')\n",
    "elif variationType == 'powerLaw':\n",
    "    fig, axes = getPaperFigureHandle('Figure S2')    \n",
    "\n",
    "ax = axes[1,2]\n",
    "ax.clear()\n",
    "plt.sca(ax)\n",
    "lineStyles = ['dotted','dashed','solid']\n",
    "lIdx = 0\n",
    "for q in quantiles:\n",
    "    means = fidelityMat[str(q)].mean(axis=2)\n",
    "    stds = fidelityMat[str(q)].std(axis=2)\n",
    "   \n",
    "\n",
    "    if variationType == 'normal':\n",
    "        if idx in [0,3,4]:\n",
    "            print( q )\n",
    "            print( means )\n",
    "            _ = plots.manyErrorBarsPlot( muExtList, means, stds, sigmaRange, cbarText='Coefficienct of variation',xIsMean=True,ls=lineStyles[lIdx],\n",
    "                            xAxisText='Mean extraction efficiency', yAxisText='Fidelity', cmap='jet', fontsize=10, plotCbar=False )    \n",
    "            lIdx += 1\n",
    "            ax.legend()\n",
    "            ax.tick_params(labelsize=10)\n",
    "    elif variationType == 'powerLaw':\n",
    "        if idx in [0,3,4]:\n",
    "            ax.errorbar( muRange, means[0], yerr=stds[0], marker='.', label=labels[idx] )\n",
    "            ax.set_xlabel('Mean extraction efficiency',fontsize=10)\n",
    "            ax.set_ylabel('Fidelity',fontsize=10)\n",
    "            ax.legend()\n",
    "    \n",
    "    ax.set_ylim([0,1])\n",
    "    ax.set_xscale('log')\n",
    "    idx += 1\n",
    "\n",
    "ax.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [Figure 2C] PCR amplification efficiency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "variationType = 'powerLaw'\n",
    "\n",
    "if variationType == 'normal':\n",
    "    muAmpList = [10,100,1000,10000]\n",
    "    sigmaRange = [0.0,0.5,1.0]\n",
    "elif variationType == 'powerLaw':\n",
    "    muAmpList = [1.0]\n",
    "    sigmaRange = [-0.90,-0.5,-0.1,0.8,4]\n",
    "#    muRange = np.zeros_like( sigmaRange )\n",
    "\n",
    "ratioColumn = 'ratio'\n",
    "\n",
    "numReps = 10\n",
    "\n",
    "quantiles = [[0,0.25],[0.25,0.5],[0.5,0.75],[0.75,1.0],[0,1.0]]\n",
    "labels = ['Bottom 25','25th - 50th percentile','50th - 75th percentile','Top 25','All regions']\n",
    "fidelityMat = {}\n",
    "for q in quantiles:\n",
    "    fidelityMat[str(q)] = np.zeros((len(muAmpList),len(sigmaRange),numReps))\n",
    "\n",
    "directMotif = MotifTable.MotifTable( 'Tye7' )\n",
    "energiesSampled = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numLocations )\n",
    "chemicalPotential = 3\n",
    "    \n",
    "muIdx = 0\n",
    "ctr = 0\n",
    "pcrCycles = 15\n",
    "pExt = distributions.truncNorm( a=0.0, b=1.0, mu=0.5, sigma=0.05, size=numLocations )\n",
    "for mu in muAmpList:\n",
    "    sigmaIdx = 0\n",
    "    for sigma in sigmaRange:\n",
    "        if variationType == 'normal':\n",
    "            if sigma > 0:\n",
    "                ampRatio = distributions.truncNorm( a=1, b=2**pcrCycles, mu=mu, sigma=sigma*mu, size=np.int(numLocations) )                \n",
    "            else:\n",
    "                ampRatio = mu\n",
    "        elif variationType == 'powerLaw':\n",
    "            pAmp = distributions.truncPowerLawRVs( sigma, 1, 100, size=numLocations, algorithm='scipy' )/100\n",
    "            ampRatio = np.power( 1 + pAmp, pcrCycles )\n",
    "\n",
    "            muRange[sigmaIdx] = np.mean( ampRatio )\n",
    "        \n",
    "        repIdx = 0\n",
    "        for rep in range(numReps):\n",
    "            print('{},{},{}'.format(muIdx,sigmaIdx,rep))\n",
    "            if ratioColumn == 'score':\n",
    "                spEnergies, sequences = directMotif.sampleFromDistribution( energiesSampled )                \n",
    "                genome = cs.performChipSeq( spEnergies=spEnergies, pExt=pExt, numCells=100000, chemicalPotential=chemicalPotential )\n",
    "            else:\n",
    "                genome = cs.performChipSeq( spEnergies=energiesSampled, pExt=pExt, numCells=100000, chemicalPotential=chemicalPotential )                \n",
    "\n",
    "                                    \n",
    "            if ratioColumn == 'score':\n",
    "                genome = genome.query('ratio > 0 & unique_control_reads > 0')\n",
    "                learnGenome = genome.query( 'ratio > {}'.format( genome['ratio'].quantile(0.9)) )\n",
    "\n",
    "                pwmDouble, pcmDouble, pfmDouble = cs.findPWM( learnGenome['sequence'].values )\n",
    "                calculatedMotif = MotifTable.MotifTable( 'Tye7', 'yeast', pwmDouble, matType='frequencies')\n",
    "                scores = calculatedMotif.getInformation( genome['sequence'].values )\n",
    "                genome.loc[:,'score'] = scores\n",
    "\n",
    "            #genome.loc[:,'ratio'] = genome.eval('unique_chip_reads/unique_control_reads').values\n",
    "                \n",
    "            fidelityValues = util.computeFidelity( genome, ratioColumn=ratioColumn )\n",
    "            qIdx = 0\n",
    "            for q in quantiles:\n",
    "                fidelityMat[str(q)][muIdx,sigmaIdx,rep] = fidelityValues[qIdx]\n",
    "                qIdx += 1\n",
    "\n",
    "            repIdx += 1\n",
    "        sigmaIdx += 1\n",
    "    muIdx += 1\n",
    "\n",
    "idx = 0\n",
    "if variationType == 'normal':\n",
    "    fig, axes = getPaperFigureHandle('Figure 2')\n",
    "elif variationType == 'powerLaw':\n",
    "    fig, axes = getPaperFigureHandle('Figure S2')    \n",
    "    \n",
    "ax1 = axes[1,1]\n",
    "#ax2 = axes[1,2]\n",
    "ax1.clear()\n",
    "plt.sca(ax1)\n",
    "lineStyles = ['dotted','dashed','solid']\n",
    "lIdx = 0\n",
    "for q in quantiles:\n",
    "    means = fidelityMat[str(q)].mean(axis=2)\n",
    "    stds = fidelityMat[str(q)].std(axis=2)\n",
    "\n",
    "    if variationType == 'normal':\n",
    "        if idx in [0,1,2,3,4]:\n",
    "            ax = ax1\n",
    "            _ = plots.manyErrorBarsPlot( muAmpList, means, stds, sigmaRange, cbarText='Coefficienct of variation',xIsMean=True,ls=lineStyles[lIdx],\n",
    "                            xAxisText='Mean amplification ratio', yAxisText='Fidelity', cmap='jet',fontsize=10, plotCbar=False )    \n",
    "            lIdx += 1\n",
    "            ax.tick_params(labelsize=10)\n",
    "    elif variationType == 'powerLaw':\n",
    "        ax1.errorbar( muRange, means[0], yerr=stds[0], marker='.', label=labels[idx] )\n",
    "        ax1.set_xlabel('Mean amplification ratio',fontsize=10)\n",
    "        ax1.set_ylabel('Fidelity',fontsize=10)\n",
    "    \n",
    "    ax1.set_ylim([0,1])\n",
    "    ax1.set_xscale('log')\n",
    "    idx += 1\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "idx = 0\n",
    "quantiles = [[0,0.25],[0.25,0.5],[0.5,0.75],[0.75,1.0],[0,1.0]]\n",
    "\n",
    "if ratioColumn == 'ratio':\n",
    "    fig, axes = getPaperFigureHandle( 'Figure 2' )\n",
    "else:\n",
    "    fig, axes = getPaperFigureHandle( 'Figure S2' )\n",
    "ax = axes[1,1]\n",
    "#ax2 = axes[1,2]\n",
    "ax.clear()\n",
    "plt.sca(ax)\n",
    "lineStyles = ['dotted','dashed','solid']\n",
    "lIdx = 0\n",
    "for q in quantiles:\n",
    "    means = fidelityMat[str(q)].mean(axis=2)\n",
    "    stds = fidelityMat[str(q)].std(axis=2)\n",
    "\n",
    "    if variationType == 'normal':\n",
    "        if idx in [0,3,4]:\n",
    "            _ = plots.manyErrorBarsPlot( muAmpList, means, stds, sigmaRange, cbarText='Coefficienct of variation',xIsMean=True,ls=lineStyles[lIdx],\n",
    "                            xAxisText='Mean amplification ratio', yAxisText='Fidelity', cmap='jet',fontsize=10, plotCbar=False )    \n",
    "            lIdx += 1\n",
    "            ax.tick_params(labelsize=10)\n",
    "    elif variationType == 'powerLaw':\n",
    "        if idx in [0,3,4]:\n",
    "            ax.errorbar( muRange, means[0], yerr=stds[0], marker='.', label=labels[idx] )\n",
    "            ax.set_xlabel('Mean amplification ratio',fontsize=10)\n",
    "            ax.set_ylabel('Fidelity',fontsize=10)\n",
    "    \n",
    "    ax.set_ylim([0,1])\n",
    "    ax.set_xscale('log')\n",
    "    idx += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [Figure 3C] Indirect binding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     32,
     37
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "numLocations = 1000\n",
    "\n",
    "pExt = distributions.truncNorm( a=0.0, b=1.0, mu=0.5, sigma=0.05, size=numLocations )\n",
    "numReps = 10\n",
    "\n",
    "energyRanges = [[0,10]]\n",
    "fracIndirectSet = [0.001,0.01,0.1,0.2,0.3,0.4,0.5]\n",
    "\n",
    "quantiles = [[0,0.25],[0.25,0.5],[0.5,0.75],[0.75,1.0],[0,1.0]]\n",
    "toPlot = [True,False,False,True,True]\n",
    "labels = ['0-25','25-50','50-75','75-100','All regions']\n",
    "fidelityMat = {}\n",
    "for q in quantiles:\n",
    "    fidelityMat[str(q)] = np.zeros((len(energyRanges),len(fracIndirectSet),numReps))\n",
    "    \n",
    "energyIdx = 0\n",
    "\n",
    "ratioColumn = 'ratio'\n",
    "chemicalPotential = 3\n",
    "secondTFchemicalPotential = 3\n",
    "spEnergies = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numLocations )\n",
    "\n",
    "for energyRange in energyRanges:\n",
    "    fracIdx = 0\n",
    "    motifInformation = np.zeros( (len( fracIndirectSet ), numReps ) )    \n",
    "\n",
    "    secondTFspEnergies = distributions.truncPowerLawRVs( 0.5, energyRange[0], energyRange[1], size=numLocations )\n",
    "\n",
    "    for fracIndirect in fracIndirectSet:\n",
    "        numIndirect = np.int( fracIndirect*numLocations )\n",
    "\n",
    "        indirectLocations = np.arange( numIndirect )\n",
    "\n",
    "        for rep in range(numReps):\n",
    "            print( '{},{},{}'.format( energyIdx+1, fracIdx+1, rep+1))\n",
    "\n",
    "            if ratioColumn == 'score':\n",
    "                directMotif = MotifTable.MotifTable( 'Tye7', 'yeast' )\n",
    "                energiesSampled = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numLocations )\n",
    "                spEnergies, sequences = directMotif.sampleFromDistribution( 'yeast', energiesSampled )\n",
    "                \n",
    "                genome = cs.performChipSeq( sequences, spEnergies=spEnergies, pExt=pExt, secondTFspEnergies=secondTFspEnergies,\n",
    "                                           indirectLocations=indirectLocations, numCells=100000, chemicalPotential=chemicalPotential,\n",
    "                                          secondTFchemicalPotential=secondTFchemicalPotential )                \n",
    "                \n",
    "                learnGenome = genome.sort_values(by='ratio',ascending=False).head(np.int(0.1*numLocations))\n",
    "      \n",
    "                pwmDouble, pcmDouble, pfmDouble = cs.findPWM( learnGenome['sequence'].values )\n",
    "                calculatedMotif = MotifTable.MotifTable( 'Tye7', 'yeast', pwmDouble, matType='frequencies')\n",
    "                scores = calculatedMotif.getInformation( genome['sequence'].values )\n",
    "                genome.loc[:,'score'] = scores \n",
    "            elif ratioColumn == 'ratio':\n",
    "                genome = cs.performChipSeq( spEnergies=spEnergies, pExt=pExt, secondTFspEnergies=secondTFspEnergies, \n",
    "                                           indirectLocations=indirectLocations, numCells=100000, chemicalPotential=chemicalPotential,\n",
    "                                          secondTFchemicalPotential=secondTFchemicalPotential )\n",
    "                \n",
    "\n",
    "            fidelityValues = util.computeFidelity( genome, ratioColumn=ratioColumn )\n",
    "            qIdx = 0\n",
    "            for q in quantiles:\n",
    "                fidelityMat[str(q)][energyIdx,fracIdx,rep] = fidelityValues[qIdx]\n",
    "                qIdx += 1\n",
    "                \n",
    "            for ax in axes.ravel():\n",
    "                ax.set_xlabel('Binding energy')\n",
    "                ax.set_ylabel('Read count ratio')\n",
    "                ax.set_yscale('log')\n",
    "        fracIdx += 1\n",
    "    energyIdx += 1\n",
    "\n",
    "if ratioColumn == 'ratio':\n",
    "    fig, axes = getPaperFigureHandle( 'Figure 3' )\n",
    "    ax = axes[1,1]\n",
    "else:\n",
    "    fig, axes = getPaperFigureHandle( 'Figure 5' )\n",
    "    ax = axes[0,1]\n",
    "\n",
    "ax.clear()\n",
    "idx = 0\n",
    "linestyles = ['solid','dashed']\n",
    "colors = ['black',cbcolors['sky blue'],'orange']\n",
    "colorIdx = 0\n",
    "incrementColor = False\n",
    "for q in quantiles:\n",
    "    lineIdx = 0\n",
    "    for energyIdx in range(len(energyRanges)):\n",
    "        if toPlot[idx]:\n",
    "            means = fidelityMat[str(q)][energyIdx].mean(axis=1)\n",
    "            stds = fidelityMat[str(q)][energyIdx].std(axis=1)\n",
    "            ax.errorbar( fracIndirectSet, means, yerr=stds, color=colors[colorIdx], ls=linestyles[lineIdx], ms=2, marker='.' )\n",
    "            ax.set_xlabel('Fraction of indirectly bound sites')\n",
    "            ax.legend()\n",
    "            ax.set_ylim(0,1)\n",
    "            incrementColor = True\n",
    "        lineIdx += 1\n",
    "    if incrementColor:\n",
    "        colorIdx += 1\n",
    "        incrementColor = False\n",
    "    idx += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "if ratioColumn == 'ratio':\n",
    "    fig, axes = getPaperFigureHandle( 'Figure 2' )\n",
    "    ax = axes[1,1]\n",
    "else:\n",
    "    fig, axes = getPaperFigureHandle( 'Figure 5' )\n",
    "    ax = axes[0,1]\n",
    "ax.clear()\n",
    "idx = 0\n",
    "linestyles = ['solid','dashed']\n",
    "colors = ['black',cbcolors['sky blue'],'orange']\n",
    "colorIdx = 0\n",
    "incrementColor = False\n",
    "for q in quantiles:\n",
    "    lineIdx = 0\n",
    "    for energyIdx in range(len(energyRanges)):\n",
    "        if toPlot[idx]:\n",
    "            means = fidelityMat[str(q)][energyIdx].mean(axis=1)\n",
    "            stds = fidelityMat[str(q)][energyIdx].std(axis=1)\n",
    "            ax.errorbar( fracIndirectSet, means, yerr=stds, capthick=2, color=colors[colorIdx], ls=linestyles[lineIdx], ms=2, label=str(energyRanges[energyIdx]), marker='.' )\n",
    "            \n",
    "            ax.set_ylim(0,1)\n",
    "            incrementColor = True\n",
    "        lineIdx += 1\n",
    "    if incrementColor:\n",
    "        colorIdx += 1\n",
    "        incrementColor = False\n",
    "    idx += 1\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [Figure 3C] Cooperative binding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [],
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fidelityMat = {}\n",
    "\n",
    "energyRanges = [[0,10]]\n",
    "labels = [\",\".join( [str(energy) for energy in bounds]) for bounds in energyRanges]\n",
    "fracCoopSet = [0.001,0.01,0.1,0.2,0.3,0.4,0.5]\n",
    "coopTf = 'Atf1'\n",
    "numReps = 10\n",
    "ratioColumn = 'ratio'\n",
    "\n",
    "quantiles = [[0,0.25],[0.25,0.5],[0.5,0.75],[0.75,1.0],[0,1.0]]\n",
    "toPlot = [True,False,False,True,True]\n",
    "labels = ['Top 25th percentile','25th - 50th percentile','50th - 75th percentile','75th - 100th percentile','All regions']\n",
    "fidelityMat = {}\n",
    "    \n",
    "pExt = distributions.truncNorm( a=0, b=1, mu=0.5, sigma=0.05, size=numLocations )\n",
    "energyIdx = 0\n",
    "intEnergies = [-6,-2]\n",
    "\n",
    "directMotif = MotifTable.MotifTable( 'Tye7','yeast' )\n",
    "energiesSampled = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numLocations )\n",
    "\n",
    "chemicalPotential = 3\n",
    "secondTFchemicalPotential = 3\n",
    "\n",
    "for q in quantiles:\n",
    "    fidelityMat[str(q)] = np.zeros((len(energyRanges),len(intEnergies),len(fracCoopSet),numReps))\n",
    "\n",
    "for energyRange in energyRanges:\n",
    "    fracIdx = 0\n",
    "    secondTFspEnergies = distributions.truncPowerLawRVs( 0.5, energyRange[0], energyRange[1], size=numLocations )    \n",
    "    \n",
    "    intEnergyIdx = 0\n",
    "    for intEnergy in intEnergies:\n",
    "        fracIdx = 0\n",
    "        for fracCoop in fracCoopSet:\n",
    "            numCoopLocations = np.int( fracCoop*numLocations )\n",
    "\n",
    "            coopLocations = np.arange( numCoopLocations )\n",
    "\n",
    "            tfIntEnergies = np.append( np.ones( numCoopLocations)*intEnergy, np.zeros( numLocations - numCoopLocations ))\n",
    "\n",
    "            for rep in range(numReps):\n",
    "                print( '{}/{},{}/{},{}/{},{}/{}'.format( energyIdx+1, len(energyRanges), intEnergyIdx+1, len( intEnergies), fracIdx+1, len(fracCoopSet), rep+1, numReps))\n",
    "                if ratioColumn == 'ratio':\n",
    "                    genome = cs.performChipSeq( spEnergies=spEnergies, pExt=pExt, \n",
    "                                           secondTFspEnergies=secondTFspEnergies, secondTFintEnergies=tfIntEnergies, \n",
    "                                           chemicalPotential=chemicalPotential, secondTFchemicalPotential=secondTFchemicalPotential, numCells=100000 )\n",
    "                \n",
    "                elif ratioColumn == 'score':\n",
    "                    spEnergies, sequences = directMotif.sampleFromDistribution( energiesSampled )                    \n",
    "                    genome = cs.performChipSeq( sequences, spEnergies=spEnergies, pExt=pExt, \n",
    "                                           secondTFspEnergies=secondTFspEnergies, secondTFintEnergies=tfIntEnergies, \n",
    "                                           chemicalPotential=chemicalPotential, secondTFchemicalPotential=secondTFchemicalPotential, numCells=100000 )\n",
    "                    \n",
    "                    learnGenome = genome.sort_values(by='ratio',ascending=False).head(np.int(0.1*numLocations))\n",
    "          \n",
    "                    pwmDouble, pcmDouble, pfmDouble = cs.findPWM( learnGenome['sequence'].values )\n",
    "                    calculatedMotif = MotifTable.MotifTable( 'Tye7', 'yeast', pwmDouble, matType='frequencies')\n",
    "                    scores = calculatedMotif.getInformation( genome['sequence'].values )\n",
    "                    genome.loc[:,'score'] = scores                \n",
    "\n",
    "                qIdx = 0\n",
    "                fidelityValues = util.computeFidelity( genome, ratioColumn=ratioColumn )\n",
    "                for q in quantiles:\n",
    "                    fidelityMat[str(q)][energyIdx,intEnergyIdx,fracIdx,rep] = fidelityValues[qIdx]\n",
    "                    qIdx += 1\n",
    "\n",
    "            fracIdx += 1\n",
    "        intEnergyIdx += 1\n",
    "    energyIdx += 1\n",
    "\n",
    "if ratioColumn == 'ratio':\n",
    "    fig, axes = getPaperFigureHandle( 'Figure 3' )\n",
    "    ax = axes[1,2]\n",
    "else:\n",
    "    fig, axes = getPaperFigureHandle( 'Figure 5' )\n",
    "    ax = axes[0,2]\n",
    "\n",
    "ax.clear()\n",
    "idx = 0\n",
    "linestyles = ['solid','dashed']\n",
    "colors = ['black',cbcolors['sky blue'],'orange']\n",
    "colorIdx = 0\n",
    "incrementColor = False\n",
    "for q in quantiles:\n",
    "    lineIdx = 0\n",
    "    for energyIdx in range(len(energyRanges)):\n",
    "        if toPlot[idx]:\n",
    "            means = fidelityMat[str(q)][energyIdx,0].mean(axis=1)\n",
    "            stds = fidelityMat[str(q)][energyIdx,0].std(axis=1)\n",
    "            ax.errorbar( fracCoopSet, means, yerr=stds, capthick=2, color=colors[colorIdx], ls=linestyles[lineIdx], ms=2, label=str(energyRanges[energyIdx]), marker='.' )\n",
    "            ax.set_xlabel('Fraction of cooperatively bound sites')\n",
    "            ax.set_ylim(0,1)\n",
    "            incrementColor = True\n",
    "        lineIdx += 1\n",
    "    if incrementColor:\n",
    "        colorIdx += 1\n",
    "        incrementColor = False\n",
    "    idx += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if ratioColumn == 'ratio':\n",
    "    fig, axes = getPaperFigureHandle( 'Figure 3' )\n",
    "    ax = axes[1,2]\n",
    "else:\n",
    "    fig, axes = getPaperFigureHandle( 'Figure 5' )\n",
    "    ax = axes[0,2]\n",
    "\n",
    "ax.clear()\n",
    "idx = 0\n",
    "linestyles = ['solid','dashed']\n",
    "colors = ['black',cbcolors['sky blue'],'orange']\n",
    "colorIdx = 0\n",
    "incrementColor = False\n",
    "for q in quantiles:\n",
    "    lineIdx = 0\n",
    "    for energyIdx in range(len(energyRanges)):\n",
    "        if toPlot[idx]:\n",
    "            means = fidelityMat[str(q)][energyIdx,0].mean(axis=1)\n",
    "            stds = fidelityMat[str(q)][energyIdx,0].std(axis=1)\n",
    "            ax.errorbar( fracCoopSet, means, yerr=stds, capthick=2, color=colors[colorIdx], ls=linestyles[lineIdx], ms=2, label=str(energyRanges[energyIdx]), marker='.' )\n",
    "            ax.set_xlabel('Fraction of cooperatively bound sites')\n",
    "            ax.set_ylim(0,1)\n",
    "            incrementColor = True\n",
    "        lineIdx += 1\n",
    "    if incrementColor:\n",
    "        colorIdx += 1\n",
    "        incrementColor = False\n",
    "    idx += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chromatin accessibility [Supplement]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Area under ROC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "alphaList = [0.1,0.5,1.0,5.0,10,100]\n",
    "numReps = 10\n",
    "\n",
    "auROCmat = np.zeros( (len(alphaList),numReps) )\n",
    "numFPlocations = 1000\n",
    "numLocations = 1000\n",
    "numLocations += numFPlocations\n",
    "\n",
    "beta = 0.5\n",
    "Emax = 10\n",
    "energiesSampled = distributions.truncPowerLawRVs( beta, 0, Emax, size=numLocations )\n",
    "lineStyleList = ['solid','dashed']\n",
    "fpMfold = 1.37\n",
    "fpBeta, fpEmax = getFPpowerLaw( beta, Emax,  fpMfold )\n",
    "fpEnergies = distributions.truncPowerLawRVs( fpBeta, 0, fpEmax, size=numFPlocations )\n",
    "energiesSampled[(numLocations-numFPlocations):] = fpEnergies\n",
    "\n",
    "fig, axes = getPaperFigureHandle( 'Figure 5' )\n",
    "ax = axes[1,0]\n",
    "axes[1,1].clear()\n",
    "ax.clear()\n",
    "plt.sca( ax )\n",
    "\n",
    "fpItr = 0\n",
    "alphaIdx = 0\n",
    "meanAccessibility = []\n",
    "chemicalPotential = 3\n",
    "stdAccessibility = []\n",
    "for alpha in alphaList:\n",
    "    accessibility = cs.sampleAccessibility( alpha, numSamples=numLocations )\n",
    "    meanAccessibility.append( np.mean( accessibility ) )\n",
    "    stdAccessibility.append( np.std( accessibility ) )\n",
    "    if alpha in [0.5,5.0,10]:\n",
    "        pd.Series( accessibility ).plot( kind='kde', ax=axes[1,1] , label='$\\\\alpha =$ {}'.format( alpha ) )\n",
    "    print( \"{}/{}\".format(  alphaIdx, len(alphaList) ) )\n",
    "    axes[1,1].set_xlim(0,1)\n",
    "    for rep in range(numReps):\n",
    "        genome = cs.performChipSeq( spEnergies=energiesSampled, numFPlocations=numFPlocations, numCells=100000, chemicalPotential=chemicalPotential, chromAccessibility=accessibility ) \n",
    "\n",
    "        auROCmat[alphaIdx,rep] = util.findAuROC( genome )\n",
    "    alphaIdx += 1\n",
    "\n",
    "means = auROCmat.mean(axis=1)\n",
    "stds = auROCmat.std(axis=1)\n",
    "ax.errorbar( meanAccessibility, means, yerr=stds, marker='o' )\n",
    "ax = plt.gca()\n",
    "ax.set_xlabel('Mean accessbility')\n",
    "ax.set_ylabel('Area under ROC')\n",
    "\n",
    "ax.set_xscale('log')\n",
    "ax.set_ylim(0,1)\n",
    "\n",
    "axes[1,1].set_xlabel('Accesibility')\n",
    "axes[1,1].legend()\n",
    "ax.grid('on',ls='dashed')\n",
    "ax.set_axisbelow(True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Estimated number of replicates needed to infer from ChIP-seq data [Figure 5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting results for number of replicates needed for binding energy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#cs.singleTFmain(prior='powerLaw',maxReplicates=4,priorParams=[0.5,0,10], makePlot=True)\n",
    "\n",
    "import pickle\n",
    "numLocations = 1000\n",
    "kdeReplicates = 1000\n",
    "distInfo = ['powerLaw', 0.5, 0, 10]\n",
    "numTFs = 1\n",
    "prior = 'powerLaw'\n",
    "chemicalPotential = 3\n",
    "dictFileName = cs.os.path.join( 'data', 'posteriorData-N{}-K{}-distInfo-{}-numTFs-{}-prior{}-mu{}.pickle'.format( numLocations, kdeReplicates, str(distInfo), numTFs, prior, chemicalPotential ) )\n",
    "file = open(dictFileName,'rb')\n",
    "dictToLoad = pickle.load( file )\n",
    "file.close()\n",
    "\n",
    "energies = dictToLoad.keys()\n",
    "ax = plt.gca()\n",
    "ax.clear()\n",
    "for atRepNumber in [1,2,3,4,5]:\n",
    "    avgIntervalWidths = []\n",
    "    avgIntervalStds = []\n",
    "    \n",
    "    for key in dictToLoad.keys():\n",
    "        energy = key\n",
    "        intervalSet = dictToLoad[key]\n",
    "        tempIntervals = []\n",
    "        for intervals in intervalSet:\n",
    "            numReps = 1\n",
    "            for interval in intervals:\n",
    "                if numReps == atRepNumber:\n",
    "                    tempIntervals.append( interval[1] - interval[0] )\n",
    "                    break\n",
    "                numReps += 1\n",
    "\n",
    "        avgIntervalWidths.append( np.mean(tempIntervals) )\n",
    "        avgIntervalStds.append( np.std(tempIntervals) )\n",
    "\n",
    "    plt.errorbar( list(energies), avgIntervalWidths, yerr=avgIntervalStds, marker='o', ms=4, label='After replicate {}'.format( atRepNumber ) )        \n",
    "    \n",
    "ax = plt.gca()\n",
    "ax.set_axisbelow(True)\n",
    "ax.plot( [2,6], [0.71,0.71], color='red', ls='dashed', linewidth=2 )\n",
    "ax.plot( [2,6], [0.24,0.24], color='red', ls='dashed', linewidth=2 )\n",
    "\n",
    "ax.legend()\n",
    "ax.set_xlabel('Binding Energy ($k_BT$)',fontsize=10)\n",
    "ax.set_ylabel('Absolute uncertainty ($k_BT$)',fontsize=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing distribution of energy differences between two single-site variants of a binding site sequence\n",
    "\n",
    "Note : To run this block of code, download binding enery matrices from http://stormo.wustl.edu/TF-BEMs/Mono-Nuc-Models.tar.gz (last accessed 03 October 2018). \n",
    "On extraction, the required energy matrices will be present in the `BEM_models/` folder. Copy all the energy matrices in this folder into the `./data/pbm_pwms/` directory. \n",
    "\n",
    "The individual BEMs are in the following format (shown here is the BEM for Aft1) ---\n",
    "\n",
    "```\n",
    "1: -0.12 0.03 -0.12 0.22\n",
    "2: -0.00 0.07 -0.20 0.14\n",
    "3: -0.08 0.03 0.04 0.00\n",
    "4: -0.07 -0.04 -0.23 0.34\n",
    "5: -0.10 0.01 -0.10 0.20\n",
    "6: -0.11 0.06 -0.08 0.13\n",
    "7: -0.21 0.04 0.12 0.06\n",
    "8: -0.53 0.35 0.01 0.17\n",
    "9: -0.73 0.39 0.18 0.17\n",
    "10: 0.86 -0.34 -0.18 -0.34\n",
    "```\n",
    "\n",
    "Since the ```loadMotif()``` function in ```MotifTable.py``` is suited for this input format, it will not work if the file formats on the URL above are changed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "tfList = [\"AFT1\", \"ALX3\", \"ALX4\", \"ARID3A\", \"ARID5A\", \"ARO80\", \"ARX\", \"ASCL2\", \"ASG1\", \"ATF1\", \"BAPX1\", \"BARHL1\", \"BARHL2\", \"BARX1\", \"BARX2\", \"BAS1\", \"BBX\", \"BCL6B\", \"BHLHB2\", \"BSX\", \"CART1\", \"CBF1\", \"CDX1\", \"CDX2\", \"CEH\", \"CEP3\", \"CHA4\", \"CPHX\", \"CRX\", \"CUP9\", \"CUTL1\", \"DBX1\", \"DBX2\", \"DLX1\", \"DLX2\", \"DLX3\", \"DLX4\", \"DLX5\", \"DMBX1\", \"DOBOX4\", \"DOBOX5\", \"DUXL\", \"E2F\", \"ECM22\", \"EGR1\", \"EHF\", \"ELF2\", \"ELF3\", \"ELF4\", \"ELF5\", \"ELK1\", \"ELK3\", \"ELK4\", \"EMX2\", \"EN1\", \"EN2\", \"EOMES\", \"ERG\", \"ESRRA\", \"ESX1\", \"ETS1\", \"ETV1\", \"ETV3\", \"ETV4\", \"ETV5\", \"ETV6\", \"EVX1\", \"EVX2\", \"FHL1\", \"FKH1\", \"FKH2\", \"FLI1\", \"FOXA2\", \"FOXJ1\", \"FOXJ3\", \"FOXK1\", \"FOXL1\", \"GABPA\", \"GAL4\", \"GAT1\", \"GAT3\", \"GAT4\", \"GATA3\", \"GATA5\", \"GATA6\", \"GBX1\", \"GBX2\", \"GCM1\", \"GCN4\", \"GLIS2\", \"GLN3\", \"GM397\", \"GM4881\", \"GM5454\", \"GMEB1\", \"GSC\", \"GSH2\", \"GSM1\", \"GZF3\", \"HAL9\", \"HBP1\", \"HDX\", \"HIC1\", \"HLH\", \"HLX1\", \"HLXB9\", \"HMBOX1\", \"HMX1\", \"HMX2\", \"HMX3\", \"HNF4A\", \"HOMEZ\", \"HOXA1\", \"HOXA10\", \"HOXA11\", \"HOXA13\", \"HOXA2\", \"HOXA3\", \"HOXA4\", \"HOXA5\", \"HOXA6\", \"HOXA7\", \"HOXA9\", \"HOXB13\", \"HOXB3\", \"HOXB4\", \"HOXB5\", \"HOXB6\", \"HOXB7\", \"HOXB8\", \"HOXB9\", \"HOXC10\", \"HOXC11\", \"HOXC12\", \"HOXC13\", \"HOXC4\", \"HOXC5\", \"HOXC6\", \"HOXC8\", \"HOXC9\", \"HOXD1\", \"HOXD10\", \"HOXD11\", \"HOXD12\", \"HOXD13\", \"HOXD3\", \"HOXD8\", \"IPF1\", \"IRC\", \"IRF3\", \"IRF4\", \"IRF5\", \"IRF6\", \"IRX2\", \"IRX3\", \"IRX4\", \"IRX5\", \"IRX6\", \"ISGF3G\", \"ISL2\", \"ISX\", \"JUNDM2\", \"KLF7\", \"LBX2\", \"LEF1\", \"LEU3\", \"LHX1\", \"LHX2\", \"LHX3\", \"LHX4\", \"LHX5\", \"LHX6\", \"LHX8\", \"LHX9\", \"LMX1A\", \"LMX1B\", \"LYS14\", \"MAFB\", \"MAFK\", \"MATALPHA2\", \"MAX\", \"MBP1\", \"MCM1\", \"MEIS1\", \"MEOX1\", \"MET32\", \"MGA1\", \"MIG1\", \"MIG2\", \"MIG3\", \"MRG1\", \"MRG2\", \"MSX1\", \"MSX2\", \"MSX3\", \"MTF1\", \"MXL\", \"MYB\", \"MYBL1\", \"MYF6\", \"NDT80\", \"NHP6A\", \"NHP6B\", \"NKX1\", \"NKX2\", \"NKX3\", \"NKX6\", \"NR2F\", \"NRG1\", \"NSY\", \"OAF1\", \"OBOX1\", \"OBOX2\", \"OBOX3\", \"OBOX5\", \"OBOX6\", \"OCT\", \"OG2X\", \"OSR1\", \"OSR2\", \"OTP\", \"OTX1\", \"OTX2\", \"PAX4\", \"PAX6\", \"PAX7\", \"PBF1\", \"PBF2\", \"PBX1\", \"PDR1\", \"PHD1\", \"PHO2\", \"PHO4\", \"PHOX2A\", \"PHOX2B\", \"PITX1\", \"PITX2\", \"PITX3\", \"PKNOX1\", \"PKNOX2\", \"PLAGL1\", \"POU1F\", \"POU2F\", \"POU3F\", \"POU4F\", \"POU6F\", \"PROP1\", \"PRRX1\", \"PRRX2\", \"PUT3\", \"RAP1\", \"RARA\", \"RAX\", \"RDR1\", \"RDS1\", \"RDS2\", \"REF\", \"RFX3\", \"RFX4\", \"RFXDC2\", \"RGT1\", \"RHOX11\", \"RHOX6\", \"RPH1\", \"RPN4\", \"RSC3\", \"RSC30\", \"RTG3\", \"RXRA\", \"SFL1\", \"SFP1\", \"SFPI1\", \"SHOX2\", \"SIP4\", \"SIX1\", \"SIX2\", \"SIX3\", \"SIX4\", \"SIX6\", \"SKN7\", \"SMAD3\", \"SMP1\", \"SOX1\", \"SOX11\", \"SOX12\", \"SOX13\", \"SOX14\", \"SOX15\", \"SOX17\", \"SOX18\", \"SOX21\", \"SOX30\", \"SOX4\", \"SOX5\", \"SOX7\", \"SOX8\", \"SP100\", \"SP4\", \"SPDEF\", \"SPIC\", \"SPT15\", \"SRD1\", \"SRF\", \"SRY\", \"STB3\", \"STP2\", \"STP4\", \"SUM1\", \"SUT2\", \"TBF1\", \"TBP\", \"TBS1\", \"TCF1\", \"TCF2\", \"TCF3\", \"TCF7\", \"TCF7L\", \"TCFAP2A\", \"TCFAP2B\", \"TCFAP2C\", \"TCFAP2E\", \"TCFE2A\", \"TEA1\", \"TEC1\", \"TGIF1\", \"TGIF2\", \"TITF1\", \"TLX2\", \"TYE7\", \"UME6\", \"UNCX4\", \"USV1\", \"VAX1\", \"VAX2\", \"VSX1\", \"XBP1\", \"YAP1\", \"YAP6\", \"YBR239C\", \"YDR520C\", \"YER130C\", \"YGR067C\", \"YKL222C\", \"YLL054C\", \"YML081W\", \"YNR063W\", \"YOX1\", \"YPR013C\", \"YPR015C\", \"YPR196W\", \"YRM1\", \"YRR1\", \"ZBTB12\", \"ZBTB3\", \"ZBTB7B\", \"ZFP105\", \"ZFP128\", \"ZFP161\", \"ZFP187\", \"ZFP281\", \"ZFP410\", \"ZFP691\", \"ZFP740\", \"ZIC1\", \"ZIC2\", \"ZIC3\", \"ZIF268\", \"ZSCAN4\"]\n",
    "\n",
    "seqToNum = {'A' : 0,'C':1,'G':2,'T':3}\n",
    "siteVariants = [0,1,2,3,0,1,2,3]\n",
    "positions = np.arange(10) #Length of PWMs from BEEML\n",
    "\n",
    "energies = [2,3,4,5,6]\n",
    "\n",
    "energyDiffDict = {}\n",
    "for energy in energies:\n",
    "    energyDiffDict[energy] = []\n",
    "    \n",
    "tfIdx = 0\n",
    "for tf in tfList:\n",
    "    motif = MotifTable.MotifTable( tf )\n",
    "    skipTf = False\n",
    "\n",
    "    print( '{}/{}'.format( tfIdx+1, len(tfList)) )\n",
    "    for energy in energies:\n",
    "        samples = np.ones(1)*energy\n",
    "        spEnergies, sequences = motif.sampleFromDistribution( samples )\n",
    "        \n",
    "        if len( sequences ) == 0:\n",
    "            skipTf = True\n",
    "            break\n",
    "            \n",
    "        intSequences = []\n",
    "        for seq in sequences:\n",
    "            intSequences.append( [seqToNum[base] for base in seq] )\n",
    "\n",
    "        mutSites = np.arange(len(sequences[0]))\n",
    "        energyDiff = 0\n",
    "        numEnergyDifferences = len( intSequences ) * motif.pwm.shape[1] * 3         \n",
    "        energyDifferences = np.zeros( numEnergyDifferences ) - 1\n",
    "        idx = 0\n",
    "        for intSeq in intSequences:\n",
    "            for (pos,site) in zip(positions,intSeq):\n",
    "                baseEnergy = motif.pwm[site,pos]\n",
    "                mutantBases = siteVariants[(site+1):(site+4)]\n",
    "                energyDifferences[idx:(idx+3)] = abs(baseEnergy - motif.pwm[mutantBases,pos])\n",
    "                idx += 3\n",
    "        \n",
    "        energyDiffDict[energy] = np.append( energyDiffDict[energy], energyDifferences )\n",
    "    tfIdx += 1\n",
    "\n",
    "for energy in energies:\n",
    "    print(\"Energy differences of single site mutants from a binding site whose energy is {}k_BT\".format( energy ) )\n",
    "    print( pd.Series( energyDiffDict[energy] ).quantile( [0.025,0.125,0.25]) )\n",
    "    print(\"------\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FASTA generation example"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a bed file containing genomic intervals from which to sample fragments. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "\n",
    "x = pybedtools.BedTool()\n",
    "testPath = '../examples/testBed.bed'\n",
    "length = 200\n",
    "numRegions = 1000\n",
    "\n",
    "y = x.random( l=length, n=numRegions, g='/home/vishakad/Data/Genomes/yeast-genome.fa.fai').sort()\n",
    "bedDf = y.to_dataframe( names=['chr','start','end','name','score','strand'] )\n",
    "bedDf.loc[:,'strand'] = '.'\n",
    "\n",
    "fragmentLength = 100\n",
    "bedDf.loc[:,'p_occ'] = distributions.truncNorm( a=0.0, b=1, mu=0.4, sigma=0.6, size=numRegions )\n",
    "bedDf.loc[:,'p_control'] = 0.4 * distributions.truncNorm( a=0.0,b=1.0, mu=0.1, sigma=0.05, size=numRegions )\n",
    "bedDf.loc[:,'summit'] = np.random.randint( length, size=numRegions ) #Peak summit\n",
    "peakNames = ['peak_' + str(peakNumber) for peakNumber in np.arange( 1, bedDf.shape[0]+1 ).astype(np.int)]\n",
    "bedDf.loc[:,'name'] = peakNames\n",
    "bedDf.loc[:,'score'] = np.int64( 1000*bedDf['p_occ'])\n",
    "\n",
    "numCells = 1000\n",
    "bedDf.loc[:,'chip_fragment_count'] = np.random.binomial( numCells, bedDf['p_occ'], size=numRegions )\n",
    "bedDf.loc[:,'control_fragment_count'] = np.random.binomial( numCells, bedDf['p_control'], size=numRegions )\n",
    "fragmentJitter = 20\n",
    "readLength = 42\n",
    "bedDf.loc[:,'thickStart'] = bedDf['summit']\n",
    "bedDf.loc[:,'thickEnd'] = bedDf['summit'] + 1\n",
    "#bedDf[['chr','start','end','name','score','strand','thickStart','thickEnd']].to_csv(testPath,sep=\"\\t\",index=False,header=False)\n",
    "\n",
    "bedDf.loc[:,'p_ext'] = 0.1\n",
    "bedDf.loc[:,'p_amp'] = 0.2\n",
    "bedDf.loc[:,'energy_A'] = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numRegions )\n",
    "bedDf[['chr','start','end','p_ext','p_amp','energy_A']].to_csv('../examples/fasta-basicExample2.tsv',sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "regionPrefix = 'region_'\n",
    "\n",
    "for (fileName,fragmentStr) in zip(['chip.bed','control.bed'],['chip_fragment_count','control_fragment_count']):\n",
    "    numFragments = bedDf[fragmentStr].sum()\n",
    "    readsDf = pd.DataFrame( )\n",
    "    readsDf.loc[:,'chr'] = np.zeros( numFragments, dtype=np.object )\n",
    "    readsDf.loc[:,'start'] = np.zeros( numFragments, dtype=np.int64 )\n",
    "    readsDf.loc[:,'end'] = np.zeros( numFragments, dtype=np.int64 )\n",
    "    readsDf.loc[:,'name'] = np.zeros( numFragments, dtype=np.object )\n",
    "    readsDf.loc[:,'score'] = np.ones( numFragments, dtype=np.int )\n",
    "    readsDf.loc[:,'strand'] = np.zeros( numFragments, dtype=np.object )\n",
    "\n",
    "    readIdx = 0\n",
    "    strand = np.zeros( numFragments, dtype=np.object )\n",
    "    strandRand = np.random.random( size=numFragments )\n",
    "    posStrand = strandRand >= 0.5\n",
    "    negStrand = ~posStrand\n",
    "    strand[posStrand] = '+'\n",
    "    strand[negStrand] = '-'\n",
    "\n",
    "    peakReadNames = np.repeat( peakNames, bedDf[fragmentStr] )\n",
    "    idx = 0\n",
    "    for (readName,count) in zip(peakNames,bedDf[fragmentStr]):\n",
    "        peakReadNames[idx:(idx+count)] = [readName + '_' + str(readNumber) for readNumber in np.arange(1,count+1).astype(np.int)]\n",
    "        idx += count\n",
    "\n",
    "    readsDf.loc[:,'strand'] = strand\n",
    "    if fragmentStr == 'chip_fragment_count':\n",
    "        fragmentMidPoints = np.repeat( bedDf['start'].values + bedDf['summit'].values, bedDf[fragmentStr])\n",
    "        fragmentStartPos = np.random.normal( fragmentMidPoints, fragmentJitter, size=numFragments ).astype(np.int64)\n",
    "    elif fragmentStr == 'control_fragment_count':\n",
    "        fragmentMidPoints = np.repeat( bedDf['start'].values, bedDf[fragmentStr] )\n",
    "        fragmentStartPos = fragmentMidPoints + length * np.random.randint( length+1, size=numFragments )\n",
    "        \n",
    "    readsDf.loc[posStrand,'start'] = fragmentStartPos[posStrand]\n",
    "    readsDf.loc[posStrand,'end'] = fragmentStartPos[posStrand] + readLength - 1\n",
    "    readsDf.loc[negStrand,'start'] = fragmentStartPos[negStrand] + fragmentLength - readLength\n",
    "    readsDf.loc[negStrand,'end'] = fragmentStartPos[negStrand] + fragmentLength - 1\n",
    "    readsDf.loc[:,'name'] = peakReadNames\n",
    "    readsDf.loc[:,'chr'] = np.repeat( bedDf['chr'].values, bedDf[fragmentStr].values )\n",
    "\n",
    "    readsDf.to_csv('../examples/{}'.format( fileName ),sep=\"\\t\",index=False,header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [Supplement] Compute K-L distance of motif due to extraction and PCR efficiency changes for different TFs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the snippet below after downloading the BEEML energy matrices into the `data/pbm_pwms/` folder. This snippet creates a CSV file `tf-energyLimits.csv` in the `data/` folder. This table contains the binding energies of the highest and lowest affinity sites of each TF in the BEEML collection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-03T09:37:04.785213Z",
     "start_time": "2019-01-03T09:37:04.643346Z"
    }
   },
   "outputs": [],
   "source": [
    "tfList = [\"AFT1\", \"ALX3\", \"ALX4\", \"ARID3A\", \"ARID5A\", \"ARO80\", \"ARX\", \"ASCL2\", \"ASG1\", \"ATF1\", \"BAPX1\", \"BARHL1\", \"BARHL2\", \"BARX1\", \"BARX2\", \"BAS1\", \"BBX\", \"BCL6B\", \"BHLHB2\", \"BSX\", \"CART1\", \"CBF1\", \"CDX1\", \"CDX2\", \"CEH\", \"CEP3\", \"CHA4\", \"CPHX\", \"CRX\", \"CUP9\", \"CUTL1\", \"DBX1\", \"DBX2\", \"DLX1\", \"DLX2\", \"DLX3\", \"DLX4\", \"DLX5\", \"DMBX1\", \"DOBOX4\", \"DOBOX5\", \"DUXL\", \"E2F\", \"ECM22\", \"EGR1\", \"EHF\", \"ELF2\", \"ELF3\", \"ELF4\", \"ELF5\", \"ELK1\", \"ELK3\", \"ELK4\", \"EMX2\", \"EN1\", \"EN2\", \"EOMES\", \"ERG\", \"ESRRA\", \"ESX1\", \"ETS1\", \"ETV1\", \"ETV3\", \"ETV4\", \"ETV5\", \"ETV6\", \"EVX1\", \"EVX2\", \"FHL1\", \"FKH1\", \"FKH2\", \"FLI1\", \"FOXA2\", \"FOXJ1\", \"FOXJ3\", \"FOXK1\", \"FOXL1\", \"GABPA\", \"GAL4\", \"GAT1\", \"GAT3\", \"GAT4\", \"GATA3\", \"GATA5\", \"GATA6\", \"GBX1\", \"GBX2\", \"GCM1\", \"GCN4\", \"GLIS2\", \"GLN3\", \"GM397\", \"GM4881\", \"GM5454\", \"GMEB1\", \"GSC\", \"GSH2\", \"GSM1\", \"GZF3\", \"HAL9\", \"HBP1\", \"HDX\", \"HIC1\", \"HLH\", \"HLX1\", \"HLXB9\", \"HMBOX1\", \"HMX1\", \"HMX2\", \"HMX3\", \"HNF4A\", \"HOMEZ\", \"HOXA1\", \"HOXA10\", \"HOXA11\", \"HOXA13\", \"HOXA2\", \"HOXA3\", \"HOXA4\", \"HOXA5\", \"HOXA6\", \"HOXA7\", \"HOXA9\", \"HOXB13\", \"HOXB3\", \"HOXB4\", \"HOXB5\", \"HOXB6\", \"HOXB7\", \"HOXB8\", \"HOXB9\", \"HOXC10\", \"HOXC11\", \"HOXC12\", \"HOXC13\", \"HOXC4\", \"HOXC5\", \"HOXC6\", \"HOXC8\", \"HOXC9\", \"HOXD1\", \"HOXD10\", \"HOXD11\", \"HOXD12\", \"HOXD13\", \"HOXD3\", \"HOXD8\", \"IPF1\", \"IRC\", \"IRF3\", \"IRF4\", \"IRF5\", \"IRF6\", \"IRX2\", \"IRX3\", \"IRX4\", \"IRX5\", \"IRX6\", \"ISGF3G\", \"ISL2\", \"ISX\", \"JUNDM2\", \"KLF7\", \"LBX2\", \"LEF1\", \"LEU3\", \"LHX1\", \"LHX2\", \"LHX3\", \"LHX4\", \"LHX5\", \"LHX6\", \"LHX8\", \"LHX9\", \"LMX1A\", \"LMX1B\", \"LYS14\", \"MAFB\", \"MAFK\", \"MATALPHA2\", \"MAX\", \"MBP1\", \"MCM1\", \"MEIS1\", \"MEOX1\", \"MET32\", \"MGA1\", \"MIG1\", \"MIG2\", \"MIG3\", \"MRG1\", \"MRG2\", \"MSX1\", \"MSX2\", \"MSX3\", \"MTF1\", \"MXL\", \"MYB\", \"MYBL1\", \"MYF6\", \"NDT80\", \"NHP6A\", \"NHP6B\", \"NKX1\", \"NKX2\", \"NKX3\", \"NKX6\", \"NR2F\", \"NRG1\", \"NSY\", \"OAF1\", \"OBOX1\", \"OBOX2\", \"OBOX3\", \"OBOX5\", \"OBOX6\", \"OCT\", \"OG2X\", \"OSR1\", \"OSR2\", \"OTP\", \"OTX1\", \"OTX2\", \"PAX4\", \"PAX6\", \"PAX7\", \"PBF1\", \"PBF2\", \"PBX1\", \"PDR1\", \"PHD1\", \"PHO2\", \"PHO4\", \"PHOX2A\", \"PHOX2B\", \"PITX1\", \"PITX2\", \"PITX3\", \"PKNOX1\", \"PKNOX2\", \"PLAGL1\", \"POU1F\", \"POU2F\", \"POU3F\", \"POU4F\", \"POU6F\", \"PROP1\", \"PRRX1\", \"PRRX2\", \"PUT3\", \"RAP1\", \"RARA\", \"RAX\", \"RDR1\", \"RDS1\", \"RDS2\", \"REF\", \"RFX3\", \"RFX4\", \"RFXDC2\", \"RGT1\", \"RHOX11\", \"RHOX6\", \"RPH1\", \"RPN4\", \"RSC3\", \"RSC30\", \"RTG3\", \"RXRA\", \"SFL1\", \"SFP1\", \"SFPI1\", \"SHOX2\", \"SIP4\", \"SIX1\", \"SIX2\", \"SIX3\", \"SIX4\", \"SIX6\", \"SKN7\", \"SMAD3\", \"SMP1\", \"SOX1\", \"SOX11\", \"SOX12\", \"SOX13\", \"SOX14\", \"SOX15\", \"SOX17\", \"SOX18\", \"SOX21\", \"SOX30\", \"SOX4\", \"SOX5\", \"SOX7\", \"SOX8\", \"SP100\", \"SP4\", \"SPDEF\", \"SPIC\", \"SPT15\", \"SRD1\", \"SRF\", \"SRY\", \"STB3\", \"STP2\", \"STP4\", \"SUM1\", \"SUT2\", \"TBF1\", \"TBP\", \"TBS1\", \"TCF1\", \"TCF2\", \"TCF3\", \"TCF7\", \"TCF7L\", \"TCFAP2A\", \"TCFAP2B\", \"TCFAP2C\", \"TCFAP2E\", \"TCFE2A\", \"TEA1\", \"TEC1\", \"TGIF1\", \"TGIF2\", \"TITF1\", \"TLX2\", \"TYE7\", \"UME6\", \"UNCX4\", \"USV1\", \"VAX1\", \"VAX2\", \"VSX1\", \"XBP1\", \"YAP1\", \"YAP6\", \"YBR239C\", \"YDR520C\", \"YER130C\", \"YGR067C\", \"YKL222C\", \"YLL054C\", \"YML081W\", \"YNR063W\", \"YOX1\", \"YPR013C\", \"YPR015C\", \"YPR196W\", \"YRM1\", \"YRR1\", \"ZBTB12\", \"ZBTB3\", \"ZBTB7B\", \"ZFP105\", \"ZFP128\", \"ZFP161\", \"ZFP187\", \"ZFP281\", \"ZFP410\", \"ZFP691\", \"ZFP740\", \"ZIC1\", \"ZIC2\", \"ZIC3\", \"ZIF268\", \"ZSCAN4\"]\n",
    "if not cs.os.path.isfile( 'data/tf-energyLimits.csv' ):\n",
    "    energiesSampled = np.arange(11)\n",
    "    numTFs = len( tfList )\n",
    "    tfDf = pd.DataFrame( {'tf' : tfList, 'minEnergy' : np.zeros(numTFs), 'maxEnergy' : np.zeros( numTFs )}, )\n",
    "    idx = 0\n",
    "    for tf in tfList:\n",
    "        motif = MotifTable.MotifTable( tf )\n",
    "        print( '{}/{}'.format( idx+1, len(tfList)))\n",
    "        spEnergies, sequences = motif.sampleFromDistribution( energiesSampled )\n",
    "        tfDf.loc[idx,'minEnergy'] = motif.minEnergy\n",
    "        tfDf.loc[idx,'maxEnergy'] = motif.maxEnergy\n",
    "        idx += 1\n",
    "    tfDf.to_csv( 'data/tf-energyLimits.csv',sep=\"\\t\",index=False)\n",
    "else:\n",
    "    tfDf = pd.read_csv( 'data/tf-energyLimits.csv',sep='\\t')\n",
    "    \n",
    "tfsSelectedDf = tfDf.query('maxEnergy >= 10')   #Choose TFs whose energies can fall in the range [0,10kBT]\n",
    "\n",
    "#Read in the transfac file containing PWM id, conserved site, TF name, and structural class. The fifth column is blank.\n",
    "transfacDf = pd.read_csv( 'PWM-con-name-class', sep=\"\\t\", header=None, names=['pwm','site','tf','domain','nan'] )\n",
    "\n",
    "#Convert lower-case to upper-case names\n",
    "transfacDf.loc[:,'tf'] = [s.upper() for s in transfacDf['tf'].values]\n",
    "\n",
    "#Choose TFs from BEEML database (stored in tfsSelectedDf)\n",
    "df = transfacDf.merge( tfsSelectedDf, on=('tf') ).drop_duplicates('tf')\n",
    "\n",
    "dfg = df.groupby('domain')\n",
    "np.random.seed(10) #Set random seed to choose the same set of TFs on each run\n",
    "domainTfDict = {}\n",
    "finalTfList = []\n",
    "for domain in dfg.groups.keys():\n",
    "    tfIndex = np.random.choice( dfg.groups[domain] )\n",
    "    chosenTf = df.loc[tfIndex,'tf']\n",
    "    domainTfDict[domain] = chosenTf\n",
    "    finalTfList.append( chosenTf )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After the above snippet is run, run the snippet below to compute the K-L distances of each TF. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-03T09:37:14.452105Z",
     "start_time": "2019-01-03T09:37:14.357488Z"
    }
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unsupported operand type(s) for +: 'int' and 'str'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-99-1257afce42ea>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     50\u001b[0m     \u001b[0menergiesSampled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdistributions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtruncPowerLawRVs\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0;36m0.5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnumLocations\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 52\u001b[0;31m \u001b[0;36m4\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     53\u001b[0m \u001b[0msaveFile\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0;34m'data/klDistInfo.pickle'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'wb'\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     54\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: unsupported operand type(s) for +: 'int' and 'str'"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "tfsSelected = finalTfList\n",
    "\n",
    "muList = {'normal' : {'pExt' : [0.001,0.01,0.1,0.5,0.9], 'ampRatio' : [10,100,1000,10000]}, \n",
    "          'powerLaw' : {'pExt' : [-0.9,-0.5,0.90,4], 'ampRatio' : [-0.90,-0.5,-0.1,0.8,4]}}\n",
    "cvList = [0,0.5,1.0]\n",
    "numReps = 10\n",
    "numLocations = 1000\n",
    "klDistInfo = {}\n",
    "\n",
    "pcrCycles = 15\n",
    "\n",
    "chemicalPotential = 3\n",
    "#vRange = np.zeros_like( cvList )\n",
    "\n",
    "pExt = {'normal' : {}, 'powerLaw' : {}}\n",
    "ampRatio = {'normal' : {}, 'powerLaw' : {}}\n",
    "for variationType in ['normal','powerLaw']:\n",
    "    for quantity in muList[variationType].keys():\n",
    "        for mu in muList[variationType][quantity]:\n",
    "            if variationType == 'powerLaw':\n",
    "                dictKey = str( mu )\n",
    "                pExt[variationType][dictKey] = distributions.truncPowerLawRVs( mu, 0, 1, size=np.int(numLocations), algorithm='scipy' )\n",
    "                pAmp = distributions.truncPowerLawRVs( mu, 1, 100, size=numLocations, algorithm='scipy' )/100\n",
    "                ampRatio[variationType][dictKey] = np.power( 1 + pAmp, pcrCycles )   \n",
    "                continue\n",
    "            \n",
    "            for width in cvList:\n",
    "                if variationType == 'normal':\n",
    "                    dictKey = '{}-{}'.format( mu, width )\n",
    "                    if width > 0:\n",
    "                        pExt[variationType][dictKey] = distributions.truncNorm( a=0, b=1, mu=mu, sigma=width*mu, size=np.int(numLocations) )\n",
    "                        ampRatio[variationType][dictKey] = distributions.truncNorm( a=1, b=2**pcrCycles, mu=mu, sigma=width*mu, size=np.int(numLocations) )\n",
    "                    else:\n",
    "                        pExt[variationType][dictKey] = mu\n",
    "                        ampRatio[variationType][dictKey] = mu\n",
    "    \n",
    "chipVariablesDict = {'pExt' : pExt, 'ampRatio' : ampRatio }\n",
    "numReps = 10\n",
    "\n",
    "defaultAmpRatio = 1000\n",
    "if cs.os.path.isfile( 'data/klDistInfo.pickle'):\n",
    "    prevFile = open( 'data/klDistInfo.pickle', 'rb')\n",
    "    klDistInfo = pickle.load( prevFile )\n",
    "    defaultPext = klDistInfo['defaultPext']\n",
    "    energiesSampled = klDistInfo['energies']\n",
    "    prevFile.close()\n",
    "else:\n",
    "    defaultPext = distributions.truncNorm( a=0, b=1, mu=0.5, sigma=0.05, size=numLocations )\n",
    "    energiesSampled = distributions.truncPowerLawRVs( 0.5, 0, 10, size=numLocations )\n",
    "\n",
    "4 + \"\"\n",
    "        \n",
    "tfIdx = 0\n",
    "klDistInfo['energies' ] = energiesSampled\n",
    "klDistInfo['defaultPext'] = defaultPext \n",
    "\n",
    "for tf in tfsSelected:\n",
    "    print( '{}/{}'.format( tfIdx+1, len( tfsSelected )))\n",
    "    if tf in klDistInfo.keys():\n",
    "        print( '{} already done.'.format( tf ) )\n",
    "        tfIdx += 1\n",
    "        continue\n",
    "        \n",
    "    baselineEnergies, baselineSequences, motifSequences, baselinePFM = cs.getBaselineMotif(chemicalPotential=3,numLocations=numLocations,tf=tf,energiesSampled = energiesSampled)\n",
    "    \n",
    "    klDistInfo[tf] = {'normal':{},'powerLaw':{}}\n",
    "    for variationType in ['powerLaw','normal']:\n",
    "        print( variationType )\n",
    "        klDistInfo[tf][variationType] = {}\n",
    "        klDistInfo[tf][variationType]['pExt'] = {'normal' : [], 'powerLaw' : []}\n",
    "        klDistInfo[tf][variationType]['ampRatio'] = {'normal' : [], 'powerLaw' : []}    \n",
    "        \n",
    "        for variable in ['pExt','ampRatio']:\n",
    "            if variationType == 'normal':\n",
    "                klDistInfo[tf][variationType][variable] = np.zeros( (len(muList[variationType][variable]), len(cvList),numReps) )\n",
    "            else:\n",
    "                klDistInfo[tf][variationType][variable] = np.zeros( (len(muList[variationType][variable]), numReps) )\n",
    "            \n",
    "     \n",
    "            muIdx = 0\n",
    "            for mu in muList[variationType][variable]:\n",
    "                if variationType == 'powerLaw':\n",
    "                    dictKey = '{}'.format( mu )\n",
    "                \n",
    "                widthIdx = 0\n",
    "                for width in cvList:\n",
    "                    print( '{} : {}/{}, {}/{}'.format( tf, muIdx+1, len(muList[variationType][variable]), widthIdx+1, len(cvList) ) )\n",
    "                    if variationType == 'normal':\n",
    "                        dictKey = '{}-{}'.format( mu, width )\n",
    "\n",
    "                    if variable == 'pExt':\n",
    "                        currentPext = pExt[variationType][dictKey]\n",
    "                        currentAmpRatio = defaultAmpRatio\n",
    "                    else:\n",
    "                        currentAmpRatio = ampRatio[variationType][dictKey]\n",
    "                        currentPext = defaultPext\n",
    "\n",
    "                    klDist = []\n",
    "                    for rep in range(numReps):\n",
    "                        genome = cs.performChipSeq( baselineSequences, baselineEnergies, pExt=currentPext,\n",
    "                                           numCells=100000, chemicalPotential=chemicalPotential, ampRatio=currentAmpRatio )\n",
    "                        \n",
    "                        genome = genome.sort_values(by='ratio',ascending=False)\n",
    "                        learnGenome = genome.head( np.int(0.1*numLocations) )\n",
    "\n",
    "                        pwmSingle, pcmSingle, pfmSingle = cs.findPWM( learnGenome['sequence'].values )\n",
    "\n",
    "                        klDist.append( MotifTable.klDist( baselinePFM, pfmSingle )/baselinePFM.shape[1] )\n",
    "                        \n",
    "                    if variationType == 'powerLaw':\n",
    "                        klDistInfo[tf]['powerLaw'][variable][muIdx,:] = klDist\n",
    "                        break\n",
    "                    else:\n",
    "                        klDistInfo[tf]['normal'][variable][muIdx,widthIdx,:] = klDist                    \n",
    "                        widthIdx += 0\n",
    "                        \n",
    "                muIdx += 1\n",
    "    saveFile = open( 'data/klDistInfo.pickle', 'wb' )\n",
    "    pickle.dump( klDistInfo, saveFile )\n",
    "    saveFile.close()\n",
    "    tfIdx += 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-03T09:37:34.086361Z",
     "start_time": "2019-01-03T09:37:34.070284Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['defaultPext', 'energies', 'RFX3'])"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "klDistInfo.keys()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "135px",
    "width": "252px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "820px",
    "left": "0px",
    "right": "1643px",
    "top": "111.133px",
    "width": "331px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
